<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<link href="data:text/css;charset=utf-8,%0A%40font%2Dface%20%7B%0Afont%2Dfamily%3A%20octicons%2Dlink%3B%0Asrc%3A%20url%28data%3Afont%2Fwoff%3Bcharset%3Dutf%2D8%3Bbase64%2Cd09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM%2B8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB%2FaFGpk3jaTY6xa8JAGMW%2FO62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v%2Bk%2F0an2i%2BitHDw3v2%2B9%2BDBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3%2FI7AtxEJLtzzuZfI%2BVVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy%2FLt7Kc%2B0vWY%2FgAgIIEqAN9we0pwKXreiMasxvabDQMM4riO%2BqxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw%2Bymhce7vwM9jSqO8JyVd5RH9gyTt2%2BJ%2FyUmYlIR0s04n6%2B7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv%2FocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi%2BW2%2BMjCzMIDApSwvXzC97Z4Ig8N%2FBxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh%2F8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT%2BAEjAwuDFpBmA9KMDEwMCh9i%2Fv8H8sH0%2F4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9%2FlqYwOGZxeUelN2U2R6%2BcArgtCJpauW7UQBqnFkUsjAY%2FkOU1cP%2BDAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl%2BvvmM%2FbyA48e6tWrKArm4ZJlCbdsrxksL1AwWn%2FyBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO%2F%2FsdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd%2F89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF%2B9JOS0nbaaYDCQfwCJ7Au3AHj%2BLO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm%2BEBXuAbHmIMSRMs%2B4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL%2BhD7C1xoaHeLJSEao0FEW14ckxC%2BTU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13%2F%2Blm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl%2B9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O%2FAdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB%2F%2F%2FAA8AAQAAAAAAAAAAAAAAAAABAAAAAA%3D%3D%29%20format%28%27woff%27%29%3B%0A%7D%0Abody%20%7B%0A%2Dwebkit%2Dtext%2Dsize%2Dadjust%3A%20100%25%3B%0Atext%2Dsize%2Dadjust%3A%20100%25%3B%0Acolor%3A%20%23333%3B%0Afont%2Dfamily%3A%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20%22Segoe%20UI%22%2C%20Arial%2C%20freesans%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0Afont%2Dsize%3A%2016px%3B%0Aline%2Dheight%3A%201%2E6%3B%0Aword%2Dwrap%3A%20break%2Dword%3B%0A%7D%0Aa%20%7B%0Abackground%2Dcolor%3A%20transparent%3B%0A%7D%0Aa%3Aactive%2C%0Aa%3Ahover%20%7B%0Aoutline%3A%200%3B%0A%7D%0Astrong%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%202em%3B%0Amargin%3A%200%2E67em%200%3B%0A%7D%0Aimg%20%7B%0Aborder%3A%200%3B%0A%7D%0Ahr%20%7B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Aheight%3A%200%3B%0A%7D%0Apre%20%7B%0Aoverflow%3A%20auto%3B%0A%7D%0Acode%2C%0Akbd%2C%0Apre%20%7B%0Afont%2Dfamily%3A%20monospace%2C%20monospace%3B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ainput%20%7B%0Acolor%3A%20inherit%3B%0Afont%3A%20inherit%3B%0Amargin%3A%200%3B%0A%7D%0Ahtml%20input%5Bdisabled%5D%20%7B%0Acursor%3A%20default%3B%0A%7D%0Ainput%20%7B%0Aline%2Dheight%3A%20normal%3B%0A%7D%0Ainput%5Btype%3D%22checkbox%22%5D%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0Apadding%3A%200%3B%0A%7D%0Atable%20%7B%0Aborder%2Dcollapse%3A%20collapse%3B%0Aborder%2Dspacing%3A%200%3B%0A%7D%0Atd%2C%0Ath%20%7B%0Apadding%3A%200%3B%0A%7D%0A%2A%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0A%7D%0Ainput%20%7B%0Afont%3A%2013px%20%2F%201%2E4%20Helvetica%2C%20arial%2C%20nimbussansl%2C%20liberationsans%2C%20freesans%2C%20clean%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%234078c0%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%2C%0Aa%3Aactive%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200%3B%0Amargin%3A%2015px%200%3B%0Aoverflow%3A%20hidden%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23ddd%3B%0A%7D%0Ahr%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ahr%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%2015px%3B%0Amargin%2Dbottom%3A%2015px%3B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%2030px%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%2021px%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%2016px%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%2014px%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%2011px%3B%0A%7D%0Ablockquote%20%7B%0Amargin%3A%200%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Aol%20ol%2C%0Aul%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Droman%3B%0A%7D%0Aul%20ul%20ol%2C%0Aul%20ol%20ol%2C%0Aol%20ul%20ol%2C%0Aol%20ol%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Dalpha%3B%0A%7D%0Add%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0Afont%3A%2012px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0A%7D%0A%2Eselect%3A%3A%2Dms%2Dexpand%20%7B%0Aopacity%3A%200%3B%0A%7D%0A%2Eocticon%20%7B%0Afont%3A%20normal%20normal%20normal%2016px%2F1%20octicons%2Dlink%3B%0Adisplay%3A%20inline%2Dblock%3B%0Atext%2Ddecoration%3A%20none%3B%0Atext%2Drendering%3A%20auto%3B%0A%2Dwebkit%2Dfont%2Dsmoothing%3A%20antialiased%3B%0A%2Dmoz%2Dosx%2Dfont%2Dsmoothing%3A%20grayscale%3B%0A%2Dwebkit%2Duser%2Dselect%3A%20none%3B%0A%2Dmoz%2Duser%2Dselect%3A%20none%3B%0A%2Dms%2Duser%2Dselect%3A%20none%3B%0Auser%2Dselect%3A%20none%3B%0A%7D%0A%2Eocticon%2Dlink%3Abefore%20%7B%0Acontent%3A%20%27%5Cf05c%27%3B%0A%7D%0A%2Emarkdown%2Dbody%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%20%21important%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%20%21important%3B%0A%7D%0Aa%3Anot%28%5Bhref%5D%29%20%7B%0Acolor%3A%20inherit%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0A%2Eanchor%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%2Dright%3A%202px%3B%0Amargin%2Dleft%3A%20%2D18px%3B%0A%7D%0A%2Eanchor%3Afocus%20%7B%0Aoutline%3A%20none%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%201em%3B%0Amargin%2Dbottom%3A%2016px%3B%0Afont%2Dweight%3A%20bold%3B%0Aline%2Dheight%3A%201%2E4%3B%0A%7D%0Ah1%20%2Eocticon%2Dlink%2C%0Ah2%20%2Eocticon%2Dlink%2C%0Ah3%20%2Eocticon%2Dlink%2C%0Ah4%20%2Eocticon%2Dlink%2C%0Ah5%20%2Eocticon%2Dlink%2C%0Ah6%20%2Eocticon%2Dlink%20%7B%0Acolor%3A%20%23000%3B%0Avertical%2Dalign%3A%20middle%3B%0Avisibility%3A%20hidden%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%2C%0Ah2%3Ahover%20%2Eanchor%2C%0Ah3%3Ahover%20%2Eanchor%2C%0Ah4%3Ahover%20%2Eanchor%2C%0Ah5%3Ahover%20%2Eanchor%2C%0Ah6%3Ahover%20%2Eanchor%20%7B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah2%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah3%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah4%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah5%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah6%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%20%7B%0Avisibility%3A%20visible%3B%0A%7D%0Ah1%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%202%2E25em%3B%0Aline%2Dheight%3A%201%2E2%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah1%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah2%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%201%2E75em%3B%0Aline%2Dheight%3A%201%2E225%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah2%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%201%2E5em%3B%0Aline%2Dheight%3A%201%2E43%3B%0A%7D%0Ah3%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%201%2E25em%3B%0A%7D%0Ah4%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ah5%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%201em%3B%0Acolor%3A%20%23777%3B%0A%7D%0Ah6%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ap%2C%0Ablockquote%2C%0Aul%2C%0Aol%2C%0Adl%2C%0Atable%2C%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%204px%3B%0Apadding%3A%200%3B%0Amargin%3A%2016px%200%3B%0Abackground%2Dcolor%3A%20%23e7e7e7%3B%0Aborder%3A%200%20none%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%2Dleft%3A%202em%3B%0A%7D%0Aul%20ul%2C%0Aul%20ol%2C%0Aol%20ol%2C%0Aol%20ul%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Ali%3Ep%20%7B%0Amargin%2Dtop%3A%2016px%3B%0A%7D%0Adl%20%7B%0Apadding%3A%200%3B%0A%7D%0Adl%20dt%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%2016px%3B%0Afont%2Dsize%3A%201em%3B%0Afont%2Dstyle%3A%20italic%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Adl%20dd%20%7B%0Apadding%3A%200%2016px%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ablockquote%20%7B%0Apadding%3A%200%2015px%3B%0Acolor%3A%20%23777%3B%0Aborder%2Dleft%3A%204px%20solid%20%23ddd%3B%0A%7D%0Ablockquote%3E%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Ablockquote%3E%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Atable%20%7B%0Adisplay%3A%20block%3B%0Awidth%3A%20100%25%3B%0Aoverflow%3A%20auto%3B%0Aword%2Dbreak%3A%20normal%3B%0Aword%2Dbreak%3A%20keep%2Dall%3B%0A%7D%0Atable%20th%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Atable%20th%2C%0Atable%20td%20%7B%0Apadding%3A%206px%2013px%3B%0Aborder%3A%201px%20solid%20%23ddd%3B%0A%7D%0Atable%20tr%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Aborder%2Dtop%3A%201px%20solid%20%23ccc%3B%0A%7D%0Atable%20tr%3Anth%2Dchild%282n%29%20%7B%0Abackground%2Dcolor%3A%20%23f8f8f8%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Abackground%2Dcolor%3A%20%23fff%3B%0A%7D%0Acode%20%7B%0Apadding%3A%200%3B%0Apadding%2Dtop%3A%200%2E2em%3B%0Apadding%2Dbottom%3A%200%2E2em%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%2085%25%3B%0Abackground%2Dcolor%3A%20rgba%280%2C0%2C0%2C0%2E04%29%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0Acode%3Abefore%2C%0Acode%3Aafter%20%7B%0Aletter%2Dspacing%3A%20%2D0%2E2em%3B%0Acontent%3A%20%22%5C00a0%22%3B%0A%7D%0Apre%3Ecode%20%7B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%20100%25%3B%0Aword%2Dbreak%3A%20normal%3B%0Awhite%2Dspace%3A%20pre%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0A%2Ehighlight%20%7B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0A%2Ehighlight%20pre%2C%0Apre%20%7B%0Apadding%3A%2016px%3B%0Aoverflow%3A%20auto%3B%0Afont%2Dsize%3A%2085%25%3B%0Aline%2Dheight%3A%201%2E45%3B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0A%2Ehighlight%20pre%20%7B%0Amargin%2Dbottom%3A%200%3B%0Aword%2Dbreak%3A%20normal%3B%0A%7D%0Apre%20%7B%0Aword%2Dwrap%3A%20normal%3B%0A%7D%0Apre%20code%20%7B%0Adisplay%3A%20inline%3B%0Amax%2Dwidth%3A%20initial%3B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Aoverflow%3A%20initial%3B%0Aline%2Dheight%3A%20inherit%3B%0Aword%2Dwrap%3A%20normal%3B%0Abackground%2Dcolor%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0Apre%20code%3Abefore%2C%0Apre%20code%3Aafter%20%7B%0Acontent%3A%20normal%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%2Dsize%3A%2011px%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Epl%2Dc%20%7B%0Acolor%3A%20%23969896%3B%0A%7D%0A%2Epl%2Dc1%2C%0A%2Epl%2Ds%20%2Epl%2Dv%20%7B%0Acolor%3A%20%230086b3%3B%0A%7D%0A%2Epl%2De%2C%0A%2Epl%2Den%20%7B%0Acolor%3A%20%23795da3%3B%0A%7D%0A%2Epl%2Ds%20%2Epl%2Ds1%2C%0A%2Epl%2Dsmi%20%7B%0Acolor%3A%20%23333%3B%0A%7D%0A%2Epl%2Dent%20%7B%0Acolor%3A%20%2363a35c%3B%0A%7D%0A%2Epl%2Dk%20%7B%0Acolor%3A%20%23a71d5d%3B%0A%7D%0A%2Epl%2Dpds%2C%0A%2Epl%2Ds%2C%0A%2Epl%2Ds%20%2Epl%2Dpse%20%2Epl%2Ds1%2C%0A%2Epl%2Dsr%2C%0A%2Epl%2Dsr%20%2Epl%2Dcce%2C%0A%2Epl%2Dsr%20%2Epl%2Dsra%2C%0A%2Epl%2Dsr%20%2Epl%2Dsre%20%7B%0Acolor%3A%20%23183691%3B%0A%7D%0A%2Epl%2Dv%20%7B%0Acolor%3A%20%23ed6a43%3B%0A%7D%0A%2Epl%2Did%20%7B%0Acolor%3A%20%23b52a1d%3B%0A%7D%0A%2Epl%2Dii%20%7B%0Abackground%2Dcolor%3A%20%23b52a1d%3B%0Acolor%3A%20%23f8f8f8%3B%0A%7D%0A%2Epl%2Dsr%20%2Epl%2Dcce%20%7B%0Acolor%3A%20%2363a35c%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dml%20%7B%0Acolor%3A%20%23693a17%3B%0A%7D%0A%2Epl%2Dmh%2C%0A%2Epl%2Dmh%20%2Epl%2Den%2C%0A%2Epl%2Dms%20%7B%0Acolor%3A%20%231d3e81%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmq%20%7B%0Acolor%3A%20%23008080%3B%0A%7D%0A%2Epl%2Dmi%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dstyle%3A%20italic%3B%0A%7D%0A%2Epl%2Dmb%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmd%20%7B%0Abackground%2Dcolor%3A%20%23ffecec%3B%0Acolor%3A%20%23bd2c00%3B%0A%7D%0A%2Epl%2Dmi1%20%7B%0Abackground%2Dcolor%3A%20%23eaffea%3B%0Acolor%3A%20%2355a532%3B%0A%7D%0A%2Epl%2Dmdr%20%7B%0Acolor%3A%20%23795da3%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmo%20%7B%0Acolor%3A%20%231d3e81%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%3A%2011px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20%7B%0Alist%2Dstyle%2Dtype%3A%20none%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%2B%2Etask%2Dlist%2Ditem%20%7B%0Amargin%2Dtop%3A%203px%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20input%20%7B%0Amargin%3A%200%200%2E35em%200%2E25em%20%2D1%2E6em%3B%0Avertical%2Dalign%3A%20middle%3B%0A%7D%0A%3Achecked%2B%2Eradio%2Dlabel%20%7B%0Az%2Dindex%3A%201%3B%0Aposition%3A%20relative%3B%0Aborder%2Dcolor%3A%20%234078c0%3B%0A%7D%0Acode%20%3E%20%2Ekw%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Edt%20%7B%20color%3A%20%23ed6a43%3B%20%7D%0Acode%20%3E%20%2Edv%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Ebn%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Efl%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Ech%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Est%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Eco%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%3E%20%2Eot%20%7B%20color%3A%20%230086b3%3B%20%7D%0Acode%20%3E%20%2Eal%20%7B%20color%3A%20%23a61717%3B%20%7D%0Acode%20%3E%20%2Efu%20%7B%20color%3A%20%2363a35c%3B%20%7D%0Acode%20%3E%20%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%0Acode%20%3E%20%2Ewa%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Ecn%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Esc%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Evs%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Ess%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Eim%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eva%20%7Bcolor%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ecf%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eop%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Ebu%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eex%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Epp%20%7B%20color%3A%20%23999999%3B%20%7D%0Acode%20%3E%20%2Eat%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Edo%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%3E%20%2Ean%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ecv%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ein%20%7B%20color%3A%20%23008080%3B%20%7D%0A" rel="stylesheet">
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>

</head>

<body>

<h1 id="r-code-for-bci-coexistence-analysis">R-Code for BCI-Coexistence Analysis</h1>
<p>Oliver Purschke<br />
21 September, 2017</p>
<h1 id="load-required-packages">Load required packages</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(knitr)
<span class="kw">library</span>(vegan)
<span class="kw">library</span>(doParallel)
<span class="kw">library</span>(dplyr)
<span class="kw">library</span>(rosalia)
<span class="kw">library</span>(mistnet)
<span class="kw">library</span>(phytools)
<span class="kw">library</span>(gridExtra)
<span class="kw">library</span>(picante)
<span class="kw">library</span>(adephylo)
<span class="kw">library</span>(ggplot2)</code></pre></div>
<h1 id="phylogeny">Phylogeny</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path.phylo &lt;-<span class="st"> &quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Data/Phylo/&quot;</span>
## load BCI phylogeny
bci.tree &lt;-<span class="st"> </span><span class="kw">read.tree</span>(<span class="kw">paste</span>(path.phylo, <span class="st">&quot;dated.tree.tre&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)) 
bci.tree.OP &lt;-<span class="st"> </span><span class="kw">read.tree</span>(<span class="kw">paste</span>(path.phylo, <span class="st">&quot;dated.tree.OP.tre&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>))

<span class="co"># [1] &quot;phyfull&quot;     465             &quot;bci.tree.node.OP&quot;  495      </span>
<span class="co"># [3] &quot;traits.abund.all&quot;         &quot;phy.308&quot;   </span></code></pre></div>
<p>Check names in phylogeny (do they correspond to nate's and joe' new nomenclature?)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bci.data$phyfull &lt;-<span class="st"> </span>bci.tree</code></pre></div>
<p>Rename one species</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bci.tree$tip.label[bci.tree$tip.label==<span class="st">&quot;Appunia_seibertii&quot;</span>] &lt;-<span class="st"> &quot;Morinda_seibertii&quot;</span></code></pre></div>
<h2 id="manually-add-missing-species-to-the-phylogeny-based-on-taxonomy">Manually add missing species to the phylogeny based on taxonomy</h2>
<p>Take Joe's list to get info about family-affiliation.</p>
<h3 id="add-to-congeners">Add to congeners</h3>
<ul>
<li>&quot;Alchornea_costaricensis&quot; # added</li>
<li>&quot;Apeiba_hybrid&quot; # added</li>
<li>&quot;Bactris_coloradonis&quot; # added, check whether it belongs to b.col or b.bar</li>
<li>&quot;Banara_guianensis&quot; # # added to genus Hasseltia (Flacourtiace)</li>
<li>&quot;Bertiera_guianensis&quot; # to be added to Rubiaceae (Borojoa, Calycophyllu, Chimarrhis,Chomelia, Cosmibuena, Coussarea, Coutarea, Elaeagia, Exostema, Faramea, Genipa (#), Guettarda (#), Hamelia, Isertia, Macrocnemum, Morinda, Palicourea, Pentagonia, Pogonopus (#), Posoqueria (#), Psychotria, Randia (##), Rosenbergiod (##), Rondeletia (#), Rudgea, Tocoyena (##), Alibertia, Alseis, Amaioua, Pittoniotis, Appunia) # added</li>
</ul>
<h3 id="add-to-most-closely-related-genus">Add to most closely related genus</h3>
<ul>
<li>&quot;Cestrum_megalophyllum&quot; # family Solanaceae: Solanum (add to Solanum), Cestrum (does Markea belong to Solanaceae, yes!)</li>
<li>&quot;Cyathea_petiolata&quot; # Cyatheaceae family not represented (!! tree fern, maybe skip?!), decided to include tree ferns (but might be kicked out later)</li>
<li>&quot;Colubrina_glandulosa&quot; # to be added to &quot;Rhamnaceae&quot; (family not represented). Needed to find most closely related family. Sister to clade that holds Ficus and Cecropia (N143) # added</li>
<li>&quot;Drypetes_standleyi&quot; # (family Euphorbiacea: Hura, Hieronyma, Mabea, Maprounea, Margaritaria, Pera, Sapium, Tetrorchidiu, Acalypha, Adelia, Alchornea, Amanoa, Croton, Drypetes) # added to Margaritina</li>
<li>&quot;Ficus_colubrinae&quot; # two groups within genus Ficus add to one of them # added</li>
<li>&quot;Ficus_pertusa&quot; # two groups within genus ficus add to one of them # added</li>
<li>&quot;Geonoma_interrupta&quot; # family Aracaceae: Oenocarpus (##), Attalea (), Socratea (), Synechanthus (), Astrocaryum (), Bactris (), Chamaedorea, Elaeis # added to Oenocarpus (Arecoideae)</li>
<li>&quot;Inga_mucuna&quot; # added</li>
<li>&quot;Koanophyllon_wetmorei&quot; # family: Asteraceae (Verbesina, Vernonanthur) # added to family: Asteraceae N326</li>
<li>&quot;Lacmellea_panamensis&quot; # family Apocynaceae: Rauvolfia, Stemmadenia, Tabernaemont, Thevetia # Aspidosperma renamed to &quot;Lacmellea_panamensis&quot;</li>
<li>&quot;Pavonia_dasypetala&quot; # search for family Malvacea (Hampea) # added</li>
<li>&quot;Lycianthes_maxonii&quot; # family Solanaceae: Solanum (add to Solanum), Cestrum (does Markea belong to Solanaceae) # added</li>
<li>&quot;Maclura_tinctoria&quot; # family Moraceae: Maquira (#), Trophis, Perebea (#), Poulsenia (#), Pseudolmedia (#), Sorocea, Brosimum, Castilla (#), Ficus, Trophis # added</li>
<li>&quot;Miconia_prasina&quot; # added</li>
<li>&quot;Ocotea_whitei&quot; # added</li>
<li>&quot;Piper_schiedeanum&quot; # added</li>
<li>&quot;Pachira_sessilis&quot; # added</li>
<li>&quot;Prioria_copaifera&quot; # family Fabaceae : Schizolobium, Senna, Tachigali, Brownea, Cassia, Copaifera (#), Dialium, Hymenaea (#) # added</li>
<li>&quot;Psychotria_hoffmannseggiana&quot;, see Sedio (2012) in J.Ecol. for Psychotria phylogeny # added to Psychotria_acuminata</li>
<li>&quot;Quararibea_asterolepis&quot; # family Bombacaceae: add to N73 (holds Ceiba and Pseudobombax) # added</li>
<li>&quot;Schefflera_morototoni&quot; # added to Dendropanax</li>
<li>&quot;Simarouba_amara&quot; # family Simaroubacea: Picramnia, Quassia (#) (both distantly related in phylogeny) # added to Quassia</li>
<li>&quot;Talisia_nervosa&quot; # added</li>
<li>&quot;Tetragastris_panamensis&quot; # family Burseraceae: Trattinnicki, Bursera, Protium # added to Protium</li>
<li>&quot;Trichospermum_galeottii&quot; # add to Apeiba # added</li>
<li>&quot;Vismia_macrophylla&quot; # added</li>
<li>&quot;Xylopia_macrantha&quot; # family Annonaceae: Anaxagorea, Annona (#), Desmopsis, Guatteria, Mosannona, Oxandra, Rollinia (#), Unonopsis</li>
<li>&quot;Xylosma_chlorantha&quot; # added</li>
</ul>
<h3 id="example-how-to-add-species-based-on-node-information">Example how to add species based on node information</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bci.tree.nodes &lt;-<span class="st"> </span><span class="kw">makeNodeLabel</span>(<span class="kw">drop.tip</span>(bci.tree,
<span class="st">&quot;Gnetum_leyboldii&quot;</span>), <span class="dt">prefix =</span> <span class="st">&quot;N&quot;</span>)

<span class="kw">write.tree</span>(bci.tree.nodes, <span class="dt">file =</span> <span class="st">&quot;bci.tree.node.tre&quot;</span>)

<span class="kw">pdf</span>(<span class="st">&quot;bci.tree.nodes.pdf&quot;</span>, <span class="dt">height =</span> <span class="dv">25</span>, <span class="dt">width =</span> <span class="dv">25</span>)
<span class="kw">plot</span>(bci.tree.nodes, <span class="dt">type =</span> <span class="st">&quot;f&quot;</span>, <span class="dt">cex =</span> .<span class="dv">6</span>, <span class="dt">show.node.label =</span> <span class="ot">TRUE</span>)
<span class="kw">dev.off</span>()</code></pre></div>
<h3 id="plotting-the-extended-tree-495-taxa">Plotting the extended tree (495 taxa)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bci.tree.node.OP &lt;-<span class="st"> </span><span class="kw">read.tree</span>(<span class="st">&quot;bci.tree.node.OP.tre&quot;</span>)
<span class="kw">pdf</span>(<span class="st">&quot;bci.tree.node.OP.pdf&quot;</span>, <span class="dt">height =</span> <span class="dv">25</span>, <span class="dt">width =</span> <span class="dv">25</span>)
<span class="kw">plot</span>(bci.tree.node.OP, <span class="dt">type =</span> <span class="st">&quot;f&quot;</span>, <span class="dt">cex =</span> .<span class="dv">6</span>, <span class="dt">show.node.label =</span> <span class="ot">TRUE</span>)
<span class="kw">dev.off</span>()</code></pre></div>
<h2 id="match-phylogeny-to-trait-data">Match phylogeny to trait data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">traits.abund &lt;-
<span class="st">    </span><span class="kw">read.table</span>(<span class="kw">paste</span>(path.traits,
                     <span class="st">&quot;Nadja_Traits/abun/demographic_traits_abun.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)

<span class="kw">rownames</span>(traits.abund)[<span class="kw">rownames</span>(traits.abund)==<span class="st">&quot;Appunia_seibertii&quot;</span>] &lt;-<span class="st"> &quot;Morinda_seibertii&quot;</span>

<span class="kw">rownames</span>(traits.abund)[<span class="kw">which</span>(<span class="kw">rownames</span>(traits.abund) %in%
<span class="st">                             </span>bci.data$bci.tree.node.OP$tip.label ==
<span class="st">                             </span><span class="ot">FALSE</span>)]</code></pre></div>
<p>Delete &quot;Cyathea_petiolata&quot;, which is a fern.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">traits.abund.wo.Cyathea &lt;-
traits.abund[<span class="kw">rownames</span>(traits.abund)!=<span class="st">&quot;Cyathea_petiolata&quot;</span>, ]
<span class="kw">rownames</span>(traits.abund.wo.Cyathea)[<span class="kw">which</span>(<span class="kw">rownames</span>(traits.abund.wo.Cyathea)
%in%<span class="st"> </span>bci.data$bci.tree.node.OP$tip.label ==<span class="st"> </span><span class="ot">FALSE</span>)]

bci.data$traits.abund.all &lt;-<span class="st"> </span>traits.abund.wo.Cyathea

mat &lt;-<span class="st"> </span><span class="kw">match.phylo.comm</span>(<span class="dt">phy=</span>bci.tree.node.OP, <span class="dt">comm=</span><span class="kw">t</span>(bci.data$traits.abund.all))
mat$phy$edge.length &lt;-<span class="st"> </span>mat$phy$edge.length<span class="fl">+.1</span>

bci.data$phy<span class="fl">.308</span> &lt;-<span class="st"> </span>mat$phy
bci.data$traits.abu<span class="fl">.308</span> &lt;-<span class="st"> </span><span class="kw">t</span>(mat$comm)

<span class="kw">save</span>(bci.data, <span class="dt">file =</span> <span class="kw">paste</span>(path.data, <span class="st">&quot;bci.data.Rdata&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>))</code></pre></div>
<h3 id="plot-added-species-with-different-tip.color">Plot added species with different tip.color</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(bci.tree.node.OP)
ss &lt;-<span class="st"> </span>bci.tree.node.OP$tip.label[<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>)]

ss &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Alchornea_costaricensis&quot;</span>, <span class="st">&quot;Apeiba_hybrid&quot;</span> , <span class="st">&quot;Bactris_coloradonis&quot;</span>,
        <span class="st">&quot;Banara_guianensis&quot;</span>, <span class="st">&quot;Bertiera_guianensis&quot;</span>, <span class="st">&quot;Cestrum_megalophyllum&quot;</span>,
        <span class="st">&quot;Colubrina_glandulosa&quot;</span>, <span class="st">&quot;Colubrina_glandulosa&quot;</span>, <span class="st">&quot;Drypetes_standleyi&quot;</span>,
        <span class="st">&quot;Ficus_colubrinae&quot;</span>, <span class="st">&quot;Ficus_pertusa&quot;</span>, <span class="st">&quot;Geonoma_interrupta&quot;</span>,
        <span class="st">&quot;Inga_mucuna&quot;</span>,<span class="st">&quot;Koanophyllon_wetmorei&quot;</span>, <span class="st">&quot;Lacmellea_panamensis&quot;</span>, <span class="st">&quot;Pavonia_dasypetala&quot;</span>,
        <span class="st">&quot;Lycianthes_maxonii&quot;</span>, <span class="st">&quot;Maclura_tinctoria&quot;</span>, <span class="st">&quot;Miconia_prasina&quot;</span> ,<span class="st">&quot;Ocotea_whitei&quot;</span>,
        <span class="st">&quot;Piper_schiedeanum&quot;</span> ,<span class="st">&quot;Pachira_sessilis&quot;</span>, <span class="st">&quot;Prioria_copaifera&quot;</span>,
        <span class="st">&quot;Psychotria_hoffmannseggiana&quot;</span>, <span class="st">&quot;Quararibea_asterolepis&quot;</span>, <span class="st">&quot;Schefflera_morototoni&quot;</span>,
        <span class="st">&quot;Simarouba_amara&quot;</span>, <span class="st">&quot;Talisia_nervosa&quot;</span>, <span class="st">&quot;Tetragastris_panamensis&quot;</span>,
        <span class="st">&quot;Trichospermum_galeottii&quot;</span>, <span class="st">&quot;Vismia_macrophylla&quot;</span>, <span class="st">&quot;Xylopia_macrantha&quot;</span>,
        <span class="st">&quot;Xylosma_chlorantha&quot;</span>)

<span class="kw">pdf</span>(<span class="st">&quot;bci.tree.308.pdf&quot;</span>, <span class="dt">height =</span> <span class="dv">25</span>, <span class="dt">width =</span> <span class="dv">25</span>)
<span class="kw">plot</span>(mat$phy, <span class="dt">type =</span> <span class="st">&quot;f&quot;</span>, <span class="dt">cex =</span> .<span class="dv">9</span>, <span class="dt">show.node.label =</span> <span class="ot">TRUE</span>,
<span class="dt">tip.color=</span><span class="kw">ifelse</span>(mat$phy$tip.label %in%<span class="st"> </span>ss, <span class="st">&quot;red&quot;</span>,<span class="st">&quot;black&quot;</span>))
<span class="kw">dev.off</span>()</code></pre></div>
<h2 id="phylogenetic-signal-in-demographic-rates">Phylogenetic signal in demographic rates</h2>
<h3 id="blombergs-k">Blomberg's <em>K</em></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MultiK&lt;-<span class="st"> </span>function(tre, traits){
    <span class="kw">require</span>(phytools)
    mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">nrow =</span> <span class="kw">dim</span>(traits)[<span class="dv">2</span>])
    <span class="kw">colnames</span>(mat) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;K&quot;</span>,<span class="st">&quot;P&quot;</span>)
    <span class="kw">rownames</span>(mat) &lt;-<span class="st"> </span><span class="kw">colnames</span>(traits)
    for (i in <span class="dv">1</span>:<span class="kw">dim</span>(traits)[<span class="dv">2</span>]){
        x &lt;-<span class="st"> </span><span class="kw">phylosig</span>(tre, traits[tre$tip.label,i], <span class="dt">method=</span><span class="st">&quot;K&quot;</span>, <span class="dt">test=</span><span class="ot">TRUE</span>,
                      <span class="dt">nsim=</span><span class="dv">1000</span>)
        mat[i,] &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">as.numeric</span>(x), <span class="dv">3</span>)
    }
    mat
}

mat2 &lt;-<span class="st"> </span><span class="kw">match.phylo.comm</span>(<span class="dt">phy=</span>mat$phy, <span class="dt">comm=</span><span class="kw">t</span>(res))
                                        <span class="co"># comm = transposed species x traits matrix</span>

multk &lt;-<span class="st"> </span><span class="kw">MultiK</span>(bci.data$phy206, bci.data$traits.pca[,<span class="kw">c</span>(<span class="dv">16</span>,<span class="dv">18</span>)])
multk
<span class="kw">write.csv</span>(multk, <span class="dt">file =</span> <span class="st">&quot;multk.csv&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">read.csv</span>(<span class="st">&quot;multk.csv&quot;</span>))</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">X</th>
<th align="right">K</th>
<th align="right">P</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">abun.rec2</td>
<td align="right">0.072</td>
<td align="right">0.344</td>
</tr>
<tr class="even">
<td align="left">rec.nr.mean2</td>
<td align="right">0.020</td>
<td align="right">0.520</td>
</tr>
<tr class="odd">
<td align="left">rec.light.mean2</td>
<td align="right">0.040</td>
<td align="right">0.030</td>
</tr>
<tr class="even">
<td align="left">growth.mean2</td>
<td align="right">0.069</td>
<td align="right">0.001</td>
</tr>
<tr class="odd">
<td align="left">growth.light.mean2</td>
<td align="right">0.037</td>
<td align="right">0.037</td>
</tr>
<tr class="even">
<td align="left">mort.mean2</td>
<td align="right">0.022</td>
<td align="right">0.395</td>
</tr>
<tr class="odd">
<td align="left">mort.light.mean2</td>
<td align="right">0.024</td>
<td align="right">0.316</td>
</tr>
</tbody>
</table>
<h3 id="pagels-λ">Pagel's <em>λ</em></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MultiLambda &lt;-<span class="st"> </span>function(tre, traits){
    <span class="kw">require</span>(phytools)
    mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">ncol =</span> <span class="dv">4</span>, <span class="dt">nrow =</span> <span class="kw">dim</span>(traits)[<span class="dv">2</span>])
    <span class="kw">colnames</span>(mat) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;lambda&quot;</span>,<span class="st">&quot;logL&quot;</span>,<span class="st">&quot;logL0&quot;</span>,<span class="st">&quot;P&quot;</span>)
    <span class="kw">rownames</span>(mat) &lt;-<span class="st"> </span><span class="kw">colnames</span>(traits)
    for (i in <span class="dv">1</span>:<span class="kw">dim</span>(traits)[<span class="dv">2</span>]){
        x &lt;-<span class="st"> </span><span class="kw">phylosig</span>(tre, traits[tre$tip.label,i], <span class="dt">method=</span><span class="st">&quot;lambda&quot;</span>, <span class="dt">test=</span><span class="ot">TRUE</span>)
        mat[i,] &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">as.numeric</span>(x), <span class="dv">3</span>)
    }
    mat
}

multlam &lt;-<span class="st"> </span><span class="kw">MultiLambda</span>(<span class="dt">phy=</span>mat2$phy, <span class="dt">comm=</span><span class="kw">t</span>(mat2$comm)[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>,<span class="dv">14</span>:<span class="dv">19</span>)])
                                        <span class="co"># comm = transposed species x traits matrix</span>
multlam
<span class="kw">write.csv</span>(multlam, <span class="dt">file =</span> <span class="st">&quot;multlam.csv&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">read.csv</span>(<span class="st">&quot;multlam.csv&quot;</span>))</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">X</th>
<th align="right">lambda</th>
<th align="right">logL</th>
<th align="right">logL0</th>
<th align="right">P</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">abun.rec2</td>
<td align="right">0.941</td>
<td align="right">-1941.671</td>
<td align="right">-1953.767</td>
<td align="right">0.000</td>
</tr>
<tr class="even">
<td align="left">rec.nr.mean2</td>
<td align="right">0.166</td>
<td align="right">-401.625</td>
<td align="right">-402.322</td>
<td align="right">0.238</td>
</tr>
<tr class="odd">
<td align="left">rec.light.mean2</td>
<td align="right">0.247</td>
<td align="right">-184.106</td>
<td align="right">-184.505</td>
<td align="right">0.371</td>
</tr>
<tr class="even">
<td align="left">growth.mean2</td>
<td align="right">0.654</td>
<td align="right">-123.385</td>
<td align="right">-135.680</td>
<td align="right">0.000</td>
</tr>
<tr class="odd">
<td align="left">growth.light.mean2</td>
<td align="right">0.461</td>
<td align="right">50.587</td>
<td align="right">43.612</td>
<td align="right">0.000</td>
</tr>
<tr class="even">
<td align="left">mort.mean2</td>
<td align="right">0.563</td>
<td align="right">-263.038</td>
<td align="right">-270.492</td>
<td align="right">0.000</td>
</tr>
<tr class="odd">
<td align="left">mort.light.mean2</td>
<td align="right">0.348</td>
<td align="right">54.061</td>
<td align="right">51.666</td>
<td align="right">0.029</td>
</tr>
</tbody>
</table>
<h3 id="blombergs-k-but-including-measurement-error">Blomberg's <em>K</em>, but including measurement error</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MultiKerror&lt;-<span class="st"> </span>function(tre, traits, error){
    <span class="kw">require</span>(phytools)
    mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">ncol =</span> <span class="dv">4</span>, <span class="dt">nrow =</span> <span class="kw">dim</span>(traits)[<span class="dv">2</span>])
    <span class="kw">colnames</span>(mat) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;K&quot;</span>,<span class="st">&quot;P&quot;</span>,<span class="st">&quot;sig2&quot;</span>,<span class="st">&quot;logL&quot;</span>)
    <span class="kw">rownames</span>(mat) &lt;-<span class="st"> </span><span class="kw">colnames</span>(traits)
    for (i in <span class="dv">1</span>:<span class="kw">dim</span>(traits)[<span class="dv">2</span>]){
        x &lt;-<span class="st"> </span><span class="kw">phylosig</span>(tre, traits[tre$tip.label,i], <span class="dt">method=</span><span class="st">&quot;K&quot;</span>, <span class="dt">se=</span>error[tre$tip.label,i],
                      <span class="dt">test=</span><span class="ot">TRUE</span>, <span class="dt">nsim=</span><span class="dv">1000</span>)
        mat[i,] &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">as.numeric</span>(x), <span class="dv">3</span>)
    }
    mat
}

multkerr &lt;-<span class="st"> </span><span class="kw">MultiKerror</span>(bci.data$phy206, bci.data$traits.pca[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>)],
                        <span class="dt">error=</span>bci.data$traits.pca[,<span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>)])

multkerr
<span class="kw">write.csv</span>(multkerr, <span class="dt">file =</span> <span class="st">&quot;multkerr.csv&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">read.csv</span>(<span class="st">&quot;multkerr.csv&quot;</span>))</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">X</th>
<th align="right">K</th>
<th align="right">P</th>
<th align="right">sig2</th>
<th align="right">logL</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">rec.nr.mean2</td>
<td align="right">0.384</td>
<td align="right">0.101</td>
<td align="right">0.051</td>
<td align="right">-427.558</td>
</tr>
<tr class="even">
<td align="left">rec.light.mean2</td>
<td align="right">0.479</td>
<td align="right">0.045</td>
<td align="right">0.001</td>
<td align="right">-272.057</td>
</tr>
<tr class="odd">
<td align="left">growth.mean2</td>
<td align="right">0.460</td>
<td align="right">0.037</td>
<td align="right">0.001</td>
<td align="right">-159.899</td>
</tr>
<tr class="even">
<td align="left">growth.light.mean2</td>
<td align="right">0.746</td>
<td align="right">0.017</td>
<td align="right">0.000</td>
<td align="right">-45.587</td>
</tr>
<tr class="odd">
<td align="left">mort.mean2</td>
<td align="right">0.464</td>
<td align="right">0.005</td>
<td align="right">0.007</td>
<td align="right">-307.206</td>
</tr>
<tr class="even">
<td align="left">mort.light.mean2</td>
<td align="right">0.720</td>
<td align="right">0.336</td>
<td align="right">0.000</td>
<td align="right">-88.420</td>
</tr>
</tbody>
</table>
<h2 id="plot-abundance-6-traits-6-measurement-error-n13-on-the-phylogeny">Plot abundance + 6 traits + 6 measurement error (n=13) on the phylogeny</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">comm2 &lt;-<span class="st"> </span><span class="kw">t</span>(mat2$comm)
comm2[,<span class="dv">1</span>] &lt;-<span class="st"> </span>comm2[,<span class="dv">1</span>]^.<span class="dv">25</span>
    
herb &lt;-<span class="st"> </span><span class="kw">phylo4d</span>(mat2$phy, comm2)

<span class="kw">pdf</span>(<span class="st">&quot;BCI.traits.phylo.pdf&quot;</span>, <span class="dt">width =</span> <span class="dv">12</span>, <span class="dt">height =</span> <span class="dv">20</span>)
<span class="co"># postscript(file = &quot;BCI.traits.phylo.eps&quot;,width = 12, height = 20,</span>
paper =<span class="st"> &quot;special&quot;</span>, onefile =<span class="st"> </span><span class="ot">FALSE</span>, horizontal =<span class="st"> </span><span class="ot">FALSE</span>, pointsize=<span class="dv">12</span><span class="er">)</span>
phytab &lt;-<span class="st"> </span><span class="kw">table.phylo4d</span>(herb, <span class="dt">treetype=</span><span class="st">&quot;phylogram&quot;</span>, <span class="dt">show.node.label=</span>F,
<span class="dt">box=</span>F, <span class="dt">ratio.tree=</span><span class="dv">1</span>/<span class="dv">3</span>, <span class="dt">font=</span><span class="dv">3</span>, <span class="dt">cex.label=</span>.<span class="dv">5</span>, <span class="dt">cex.symbol=</span><span class="fl">1.2</span>,
<span class="dt">cex.legend =</span> .<span class="dv">8</span>)
<span class="kw">dev.off</span>()</code></pre></div>
<h2 id="nodewise-phylogenetic-signal-divergence-size">Nodewise phylogenetic signal (divergence size)</h2>
<p>This test has been carried out in Phylocom in the bash terminal. For interpretation of results, see Appendix in Purschke et al. (2013) J.Ecol..</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">div &lt;-<span class="st"> </span><span class="kw">xyplot</span>(Tsd.rankLow ~<span class="st"> </span>age, <span class="dt">groups=</span>trait.name, <span class="dt">data =</span>
divergence.bci, <span class="dt">type =</span> <span class="st">&quot;smooth&quot;</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">135</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(-<span class="dv">25</span>,
<span class="dv">1050</span>), <span class="dt">lty =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>), <span class="dt">par.settings =</span> <span class="kw">list</span>(<span class="dt">axis.line =</span>
<span class="kw">list</span>(<span class="dt">col =</span> <span class="dv">0</span>)),<span class="dt">scales=</span><span class="kw">list</span>(<span class="dt">col=</span><span class="dv">1</span>,<span class="dt">tck=</span><span class="kw">c</span>(-<span class="dv">1</span>,<span class="dv">0</span>)),  <span class="co"># remove top and right</span>
axes
              <span class="dt">panel=</span>function(...){
                lims &lt;-<span class="st"> </span><span class="kw">current.panel.limits</span>()
                <span class="kw">panel.xyplot</span>(...)
                <span class="kw">panel.abline</span>(<span class="dt">h=</span>lims$ylim[<span class="dv">1</span>],<span class="dt">v=</span>lims$xlim[<span class="dv">1</span>], <span class="dt">lwd =</span> <span class="fl">2.5</span>)
              },
              <span class="dt">layout.heights=</span><span class="kw">list</span>(<span class="dt">axis.xlab.padding =</span> <span class="dv">1</span>),
<span class="dt">lwd =</span> <span class="fl">1.5</span>, <span class="dt">col =</span>
               <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>,<span class="st">&quot;black&quot;</span>,<span class="st">&quot;black&quot;</span>,<span class="st">&quot;black&quot;</span>,<span class="st">&quot;black&quot;</span>,<span class="st">&quot;darkgrey&quot;</span>,<span class="st">&quot;darkgrey&quot;</span>,<span class="st">&quot;darkgrey&quot;</span>,<span class="st">&quot;darkgrey&quot;</span>,
                 <span class="st">&quot;darkgrey&quot;</span>),
<span class="dt">xlab =</span> <span class="st">&quot;Time (Myr)&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Observed divergence size (Rank)&quot;</span>,
<span class="dt">key=</span><span class="kw">list</span>(<span class="dt">space=</span><span class="st">&quot;inside&quot;</span>,  <span class="dt">between =</span> <span class="dv">1</span>, <span class="dt">padding.text =</span> <span class="dv">2</span>, <span class="dt">just =</span> <span class="kw">c</span>(.<span class="dv">7</span>,
.<span class="dv">5</span>), <span class="dt">columns =</span> <span class="dv">2</span>, <span class="dt">lines =</span> <span class="kw">list</span>(<span class="dt">lty =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>), <span class="dt">lwd =</span>
<span class="fl">1.5</span>, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>,<span class="st">&quot;black&quot;</span>,<span class="st">&quot;black&quot;</span>,<span class="st">&quot;black&quot;</span>,<span class="st">&quot;black&quot;</span>,<span class="st">&quot;darkgrey&quot;</span>,<span class="st">&quot;darkgrey&quot;</span>,<span class="st">&quot;darkgrey&quot;</span>,<span class="st">&quot;darkgrey&quot;</span>,
           <span class="st">&quot;darkgrey&quot;</span>)),text
=<span class="st"> </span><span class="kw">list</span>(<span class="kw">levels</span>(divergence.bci$trait.name))))
<span class="kw">plot</span>(div)

<span class="kw">pdf</span>(<span class="dt">file =</span> <span class="st">&quot;BCI.DivergenceSize.pdf&quot;</span>,<span class="dt">width =</span> <span class="dv">6</span>, <span class="dt">height =</span> <span class="fl">6.5</span>,
<span class="dt">pointsize=</span><span class="dv">12</span>)
<span class="co"># postscript(file = &quot;BCI.DivergenceSize.eps&quot;,width = 6, height = 6.5,</span>
paper =<span class="st"> &quot;special&quot;</span>, onefile =<span class="st"> </span><span class="ot">FALSE</span>, horizontal =<span class="st"> </span><span class="ot">FALSE</span>, pointsize=<span class="dv">12</span><span class="er">)</span>
<span class="kw">plot</span>(div)
<span class="kw">dev.off</span>()</code></pre></div>
<h1 id="new-markov-network-analyses-august-2016">New Markov-network analyses (August 2016)</h1>
<h2 id="read-in-all-environmental-data">Read in all environmental data</h2>
<p>If there are &quot;.&quot; in the name, run the following script in the bash-shell to rename the files accordingly:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">for</span> <span class="kw">filename</span> in *_txt<span class="kw">;</span> <span class="kw">do</span> <span class="ot">newname=</span><span class="kw">`echo</span> <span class="ot">$filename</span> <span class="kw">|</span> <span class="kw">sed</span> <span class="st">'s/\_txt$/.txt/g'</span><span class="kw">`</span>;
<span class="kw">mv</span> <span class="ot">$filename</span> <span class="ot">$newname</span><span class="kw">;</span> <span class="kw">done</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path &lt;-<span class="st"> &quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Data/Environment/</span>
<span class="st">Kupers_grids_bci-2016-06-27&quot;</span>
files &lt;-<span class="st"> </span><span class="kw">list.files</span>(path, <span class="dt">pattern=</span><span class="st">&quot;*.txt&quot;</span>)
for (i in <span class="dv">1</span>:<span class="kw">length</span>(files))
    <span class="kw">assign</span>(files[i], <span class="kw">read.table</span>(<span class="kw">file.path</span>(path, files[i])))</code></pre></div>
<h4 id="bring-environmental-data-in-the-same-order-as-the-species-data">Bring environmental data in the same order as the species data:</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">env.to.spec &lt;-<span class="st"> </span>function(x, ncol){
    <span class="kw">as.vector</span>(<span class="kw">matrix</span>(x, <span class="dt">ncol =</span> ncol, <span class="dt">byrow =</span> T))
}</code></pre></div>
<h4 id="load-all-env.files-for-a-particular-scale-and-put-them-into-one-dataframe">Load all env.files for a particular scale and put them into one dataframe</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(foreach)
env.scale &lt;-<span class="st"> </span><span class="kw">list</span>()

for (j in <span class="kw">c</span>(<span class="st">&quot;5m&quot;</span>,<span class="st">&quot;10m&quot;</span>,<span class="st">&quot;20m&quot;</span>,<span class="st">&quot;40m&quot;</span>,<span class="st">&quot;60m&quot;</span>,<span class="st">&quot;80m&quot;</span>,<span class="st">&quot;100m&quot;</span>)){
    filenames &lt;-<span class="st"> </span><span class="kw">list.files</span>(path, <span class="dt">pattern=</span>j, <span class="dt">full.names=</span><span class="ot">TRUE</span>)
    ldf &lt;-<span class="st"> </span><span class="kw">lapply</span>(filenames, read.table)
    res &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span>:<span class="kw">length</span>(filenames), <span class="dt">.combine =</span> cbind) %dopar%<span class="st"> </span>{
        <span class="kw">as.vector</span>(ldf[[i]])
    }
    env.scale &lt;-<span class="st"> </span><span class="kw">c</span>(env.scale, <span class="kw">list</span>(res))
}

<span class="kw">str</span>(env.scale)
<span class="kw">names</span>(env.scale) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;5m&quot;</span>,<span class="st">&quot;10m&quot;</span>,<span class="st">&quot;20m&quot;</span>,<span class="st">&quot;40m&quot;</span>,<span class="st">&quot;60m&quot;</span>,<span class="st">&quot;80m&quot;</span>,<span class="st">&quot;100m&quot;</span>)</code></pre></div>
<p>Make sure data are in the same order as species data (check for slope, ph and light):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(env.scale[[<span class="dv">7</span>]])</code></pre></div>
<p>All grids are 1000x500, apart from:</p>
<ul>
<li>40m: 1000 x 480</li>
<li>60m: 960 x 480</li>
<li>80m: 960 x 480</li>
</ul>
<p>Set up ncol-values for the <code>env.to.spec</code>-function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncol.vec &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">200</span>,<span class="dv">100</span>,<span class="dv">50</span>,<span class="dv">25</span>,<span class="dv">16</span>,<span class="dv">12</span>,<span class="dv">10</span>)

env.scale<span class="fl">.2</span> &lt;-<span class="st"> </span>env.scale

for(i in <span class="dv">1</span>:<span class="dv">7</span>){
    for (j in <span class="dv">1</span>:<span class="kw">dim</span>(env.scale[[i]])[<span class="dv">2</span>]){
        env.scale<span class="fl">.2</span>[[i]][,j] &lt;-<span class="st"> </span><span class="kw">env.to.spec</span>(env.scale[[i]][,j], <span class="dt">ncol =</span> ncol.vec[i])
    }
}

<span class="kw">str</span>(env.scale<span class="fl">.2</span>)

env.scale.correct &lt;-<span class="st"> </span>env.scale
<span class="kw">save</span>(env.scale.correct, <span class="dt">file =</span> <span class="st">&quot;env.scale.correct.Rdata&quot;</span>)</code></pre></div>
<h3 id="explore-whether-single-environmental-variables-and-abundance-grids-match-up">Explore whether single environmental variables and abundance grids match up</h3>
<h4 id="environmental-variable">Environmental variable</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filled.contour</span>(<span class="kw">matrix</span>(env.scale[[<span class="dv">3</span>]][,<span class="dv">55</span>], <span class="dt">ncol =</span> <span class="dv">25</span>, <span class="dt">byrow =</span> T),
               <span class="dt">color.palette =</span> terrain.colors, <span class="dt">main =</span> <span class="st">&quot;slope.20m&quot;</span>)

grid<span class="fl">.213.7</span> &lt;-<span class="st"> </span>bci.data$grid<span class="fl">.213</span>[<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">8</span>,<span class="dv">12</span>,<span class="dv">16</span>,<span class="dv">20</span>)]
<span class="kw">length</span>(grid<span class="fl">.213.7</span>)
<span class="kw">names</span>(grid<span class="fl">.213.7</span>)
<span class="kw">lapply</span>(grid<span class="fl">.213.7</span>, <span class="dt">FUN=</span>dim)
<span class="kw">save</span>(grid<span class="fl">.213.7</span>, <span class="dt">file =</span> <span class="st">&quot;grid.213.7.Rdata&quot;</span>)

<span class="kw">filled.contour</span>(<span class="kw">log</span>(<span class="kw">matrix</span>(grid<span class="fl">.213.7</span>[[<span class="dv">3</span>]][,<span class="dv">147</span>], <span class="dt">ncol =</span> <span class="dv">25</span>, <span class="dt">byrow =</span> T)),
               <span class="dt">color.palette =</span> terrain.colors, <span class="dt">main =</span> <span class="st">&quot;Slope.20m&quot;</span>) </code></pre></div>
<h2 id="plot-environmental-variables">Plot environmental variables</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span><span class="kw">expand.grid</span>(<span class="dt">x =</span> <span class="dv">0</span>:<span class="dv">99</span>, <span class="dt">y =</span> <span class="dv">0</span>:<span class="dv">49</span>) <span class="co"># initialize coordinates </span>
df$z &lt;-<span class="st"> </span>env.scale[[<span class="dv">2</span>]][,<span class="dv">47</span>] <span class="co"># set up fill values</span>
<span class="co"># default is compatible with geom_tile()</span>
<span class="kw">ggplot</span>(df, <span class="kw">aes</span>(x, y, <span class="dt">fill =</span> z)) +<span class="st"> </span><span class="kw">geom_raster</span>()</code></pre></div>
<p>Using <code>ggplot</code> in loop (only works for one page), e.g. at 100m scale (grain size):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p =<span class="st"> </span><span class="kw">list</span>()
df &lt;-<span class="st"> </span><span class="kw">expand.grid</span>(<span class="dt">x =</span> <span class="dv">0</span>:<span class="dv">49</span>, <span class="dt">y =</span> <span class="dv">0</span>:<span class="dv">24</span>) <span class="co"># adjust for other scales</span>

for(i in <span class="dv">1</span>:<span class="dv">28</span>) {
  df$z =<span class="st"> </span><span class="kw">env.to.spec</span>(env.scale[[<span class="dv">3</span>]][,i], <span class="dt">ncol =</span> <span class="dv">50</span>) <span class="co"># adjust ncol for other scales </span>
  p[[i]] =<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(x, y, <span class="dt">fill =</span> z)) +<span class="st"> </span>
<span class="st">      </span><span class="kw">geom_raster</span>() +<span class="st"> </span>
<span class="st">      </span><span class="kw">ggtitle</span>(<span class="kw">names</span>(env.scale[[<span class="dv">3</span>]])[i]) +<span class="st"> </span>
<span class="st">      </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colours =</span> <span class="kw">terrain.colors</span>(<span class="dv">10</span>)) +<span class="st"> </span>
<span class="st">      </span><span class="kw">theme</span>(<span class="dt">axis.title.x=</span><span class="kw">element_blank</span>(), 
        <span class="dt">axis.text.x=</span><span class="kw">element_blank</span>(), 
        <span class="dt">axis.title.y=</span><span class="kw">element_blank</span>(),
        <span class="dt">axis.text.y=</span><span class="kw">element_blank</span>(),
        <span class="dt">axis.ticks=</span><span class="kw">element_blank</span>(),
        <span class="dt">axis.line =</span> <span class="kw">element_blank</span>(),
        <span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>)+<span class="st"> </span>
<span class="st">      </span><span class="kw">labs</span>(<span class="dt">fill=</span><span class="st">&quot;&quot;</span>) <span class="co"># remove legend title</span>
}
 
<span class="kw">ggsave</span>(<span class="dt">file=</span><span class="st">&quot;env.100m.maps1.new.pdf&quot;</span>, <span class="kw">do.call</span>(<span class="st">&quot;grid.arrange&quot;</span>, <span class="kw">c</span>(p[<span class="dv">1</span>:<span class="dv">28</span>], <span class="dt">ncol=</span><span class="dv">4</span>)),
       <span class="dt">width =</span> <span class="dv">45</span>, <span class="dt">height =</span> <span class="dv">35</span>, <span class="dt">units =</span> <span class="st">&quot;cm&quot;</span>)</code></pre></div>
<h2 id="explore-env.data-pairwise-corr.-plots-pca-e.g.-at-100m-scale">Explore env.data (pairwise corr. plots, PCA), e.g. at 100m scale</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pdf</span>(<span class="st">&quot;pca.env.correct.100.pdf&quot;</span>, <span class="dt">width=</span><span class="fl">7.5</span>, <span class="dt">height=</span><span class="dv">8</span>)
dat &lt;-<span class="st"> </span>env.scale.correct.mean[[<span class="dv">7</span>]]
<span class="kw">colnames</span>(dat) &lt;-<span class="st"> </span><span class="kw">substr</span>(<span class="kw">colnames</span>(dat),<span class="dv">1</span>, <span class="dv">5</span>)
rda<span class="fl">.60</span> &lt;-<span class="st"> </span><span class="kw">rda</span>(dat, <span class="dt">scale =</span> T)
<span class="kw">plot</span>(rda<span class="fl">.60</span>, <span class="dt">type =</span> <span class="st">&quot;none&quot;</span>, <span class="dt">main=</span><span class="st">&quot;100m&quot;</span>, <span class="dt">scaling =</span> <span class="dv">3</span>)
<span class="kw">ordipointlabel</span>(rda<span class="fl">.60</span>, <span class="dt">display =</span> <span class="st">&quot;species&quot;</span>, <span class="dt">scaling =</span> <span class="dv">3</span>, <span class="dt">add =</span> <span class="ot">TRUE</span>)
<span class="kw">dev.off</span>()

<span class="kw">pdf</span>(<span class="st">&quot;cor100.correct.pdf&quot;</span>, <span class="dt">width=</span><span class="dv">25</span>, <span class="dt">height=</span><span class="dv">25</span>)
<span class="kw">CorTestPlot</span>(dat)
<span class="kw">dev.off</span>()</code></pre></div>
<p>Example for environmental PCA at the 20m scale. Covariation among predictors doesnt change much across scales.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lapply</span>(env.scale.correct, names)</code></pre></div>
<p>There are different variables in the data sets for the different scales</p>
<p>Extract only those variables that are contained at the smallest spatial scale Extract before the (\.) any character (.) any number of times (*) until the end of the string</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">..*$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">names</span>(env.scale.correct[[<span class="dv">1</span>]]))</code></pre></div>
<p>Kick out variables that do not occur in the 5m scale data set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">which</span>(<span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">..*$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">names</span>(env.scale.correct[[<span class="dv">4</span>]])) %in%<span class="st"> </span>ind)

env.scale.correct<span class="fl">.2</span> &lt;-<span class="st"> </span>env.scale.correct
for (i in <span class="dv">1</span>:<span class="dv">7</span>){
    env.scale.correct<span class="fl">.2</span>[[i]] &lt;-
<span class="st">        </span>env.scale.correct[[i]][, <span class="kw">which</span>(<span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">..*$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">names</span>(env.scale.correct[[i]])) %in%<span class="st"> </span>ind)]
}

<span class="kw">names</span>(env.scale.correct<span class="fl">.2</span>)</code></pre></div>
<p>Select soil types:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind.soil.type &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ava&quot;</span>,<span class="st">&quot;fairchild&quot;</span>,<span class="st">&quot;marron&quot;</span>,<span class="st">&quot;swamp&quot;</span>)</code></pre></div>
<p>Delete soil type columns (which occur throughout the data) to data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">env.scale.correct<span class="fl">.3</span> &lt;-<span class="st"> </span>env.scale.correct<span class="fl">.2</span>

for (i in <span class="dv">1</span>:<span class="dv">7</span>){
    env.scale.correct<span class="fl">.3</span>[[i]] &lt;-
<span class="st">        </span>env.scale.correct<span class="fl">.2</span>[[i]][, -<span class="kw">which</span>(<span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">..*$&quot;</span>, <span class="st">&quot;&quot;</span>,
                                              <span class="kw">names</span>(env.scale.correct<span class="fl">.2</span>[[i]])) %in%<span class="st"> </span>ind.soil.type)]
}

<span class="kw">lapply</span>(env.scale.correct<span class="fl">.3</span>, names)</code></pre></div>
<p>Append soil type:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">env.scale.correct<span class="fl">.4</span> &lt;-<span class="st"> </span>env.scale.correct<span class="fl">.3</span>
for (i in <span class="dv">1</span>:<span class="dv">7</span>){
    env.scale.correct<span class="fl">.4</span>[[i]] &lt;-
<span class="st">        </span><span class="kw">cbind</span>(env.scale.correct<span class="fl">.3</span>[[i]], env.scale.correct<span class="fl">.2</span>[[i]]
              [, <span class="kw">which</span>(<span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">..*$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">names</span>(env.scale.correct<span class="fl">.2</span>[[i]])) %in%<span class="st"> </span>ind.soil.type)])
}

<span class="kw">lapply</span>(env.scale.correct<span class="fl">.4</span>, names)</code></pre></div>
<h2 id="extract-the-mean-columnms-and-select-four-sets-of-predictors-at-each-spatial-scale">Extract the mean columnms and select four sets of predictors at each spatial scale:</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">env.scale.correct.mean &lt;-<span class="st"> </span>env.scale.correct<span class="fl">.4</span>
for (i in <span class="dv">2</span>:<span class="dv">7</span>){
    env.scale.correct.mean[[i]] &lt;-<span class="st"> </span>env.scale.correct.mean[[i]][,<span class="kw">c</span>(<span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">56</span>, <span class="dv">2</span>), <span class="dv">57</span>:<span class="dv">60</span>)]
}
<span class="kw">names</span>(env.scale.correct.mean)
<span class="kw">lapply</span>(env.scale.correct.mean, names)</code></pre></div>
<h3 id="st-and-2nd-pca-axis-on-all-predictors-27-vars-without-shade">1st and 2nd PCA axis on all predictors (27 vars without shade)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">env.scale<span class="fl">.2</span>.mean.PC &lt;-<span class="st"> </span>env.scale.correct.mean
<span class="kw">names</span>(env.scale<span class="fl">.2</span>.mean.PC[[<span class="dv">2</span>]])[-<span class="dv">26</span>] <span class="co"># exclude shade</span>

pca.test &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">scale</span>(env.scale.correct.mean[[<span class="dv">7</span>]][ ,-<span class="dv">26</span>]))
<span class="kw">summary</span>(pca.test)
<span class="kw">rownames</span>(pca.test$rotation)

for (i in <span class="dv">1</span>:<span class="dv">7</span>){
    env.scale<span class="fl">.2</span>.mean.PC[[i]] &lt;-
<span class="st">        </span><span class="kw">cbind</span>(env.scale.correct.mean[[i]],
              <span class="kw">scores</span>(<span class="kw">prcomp</span>(<span class="kw">scale</span>(env.scale.correct.mean[[i]][ ,-<span class="dv">26</span>])), <span class="dt">choices=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)))
}

<span class="kw">dim</span>(env.scale<span class="fl">.2</span>.mean.PC[[<span class="dv">4</span>]])

<span class="kw">tail</span>(env.scale<span class="fl">.2</span>.mean.PC[[<span class="dv">3</span>]][,<span class="dv">33</span>:<span class="dv">34</span>])</code></pre></div>
<h3 id="just-pc1-without-shade">Just PC1 (without shade)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tail</span>(env.scale<span class="fl">.2</span>.mean.PC[[<span class="dv">3</span>]][,<span class="dv">33</span>])</code></pre></div>
<h3 id="just-ph">Just pH</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(env.scale<span class="fl">.2</span>.mean.PC[[<span class="dv">3</span>]][<span class="dv">24</span>])</code></pre></div>
<h3 id="just-shade">Just shade</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(env.scale<span class="fl">.2</span>.mean.PC[[<span class="dv">3</span>]][<span class="dv">26</span>])

env.scale<span class="fl">.2</span>.mean.PC.correct &lt;-<span class="st"> </span>env.scale<span class="fl">.2</span>.mean.PC
<span class="kw">save</span>(env.scale<span class="fl">.2</span>.mean.PC.correct, <span class="dt">file =</span> <span class="st">&quot;env.scale.2.mean.PC.correct.Rdata&quot;</span>)</code></pre></div>
<h3 id="generate-reduced-data-set-with-just-5-columns">Generate reduced data set with just 5 columns</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">env.scale.small &lt;-<span class="st"> </span>env.scale<span class="fl">.2</span>.mean.PC.correct

for (i in <span class="dv">1</span>:<span class="dv">7</span>){
    env.scale.small[[i]] &lt;-<span class="st"> </span>env.scale<span class="fl">.2</span>.mean.PC.correct[[i]][,<span class="kw">c</span>(<span class="dv">24</span>,<span class="dv">8</span>,<span class="dv">26</span>,<span class="dv">33</span>,<span class="dv">34</span>)]
}
<span class="kw">lapply</span>(env.scale.small, names)

<span class="kw">save</span>(env.scale.small, <span class="dt">file =</span> <span class="st">&quot;env.scale.small.Rdata&quot;</span>)</code></pre></div>
<h2 id="generate-dataset-for-standard-dev.-of-env.-variables">Generate dataset for standard dev. of env. variables</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lapply</span>(env.scale.correct<span class="fl">.4</span>, names)
env.scale.correct.sd &lt;-<span class="st"> </span>env.scale.correct<span class="fl">.4</span>

for (i in <span class="dv">2</span>:<span class="dv">7</span>){
    env.scale.correct.sd[[i]] &lt;-<span class="st"> </span>env.scale.correct<span class="fl">.4</span>[[i]][,<span class="kw">seq</span>(<span class="dv">2</span>, <span class="dv">56</span>, <span class="dv">2</span>)]
}
<span class="kw">names</span>(env.scale.correct.sd)

<span class="kw">lapply</span>(env.scale.correct.sd, names)

env.scale.correct.sd &lt;-<span class="st"> </span>env.scale.correct.sd[-<span class="dv">1</span>]

<span class="co"># 1st and 2nd PCA axis on all predictors (27 vars without shade)</span>
env.scale<span class="fl">.2</span>.sd.PC &lt;-<span class="st"> </span>env.scale.correct.sd

<span class="kw">names</span>(env.scale<span class="fl">.2</span>.sd.PC[[<span class="dv">2</span>]])[-<span class="dv">26</span>] <span class="co"># exclude shade</span></code></pre></div>
<h3 id="explore-standard-deviation-sd-of-env.data-pairwise-corr.-plots-pca">Explore standard deviation (sd) of env.data (pairwise corr. plots, PCA)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pdf</span>(<span class="st">&quot;pca.env.correct.sd.10.pdf&quot;</span>, <span class="dt">width=</span><span class="fl">7.5</span>, <span class="dt">height=</span><span class="dv">8</span>)
dat &lt;-<span class="st"> </span>env.scale.correct.sd[[<span class="dv">1</span>]]
<span class="kw">colnames</span>(dat) &lt;-<span class="st"> </span><span class="kw">substr</span>(<span class="kw">colnames</span>(dat),<span class="dv">1</span>, <span class="dv">5</span>)
rda<span class="fl">.60</span> &lt;-<span class="st"> </span><span class="kw">rda</span>(dat, <span class="dt">scale =</span> T)
<span class="kw">plot</span>(rda<span class="fl">.60</span>, <span class="dt">type =</span> <span class="st">&quot;none&quot;</span>, <span class="dt">main=</span><span class="st">&quot;10m&quot;</span>, <span class="dt">scaling =</span> <span class="dv">3</span>)
<span class="kw">ordipointlabel</span>(rda<span class="fl">.60</span>, <span class="dt">display =</span> <span class="st">&quot;species&quot;</span>, <span class="dt">scaling =</span> <span class="dv">3</span>, <span class="dt">add =</span> <span class="ot">TRUE</span>)
<span class="kw">dev.off</span>()

<span class="kw">pdf</span>(<span class="st">&quot;cor10.correct.sd.pdf&quot;</span>, <span class="dt">width=</span><span class="dv">25</span>, <span class="dt">height=</span><span class="dv">25</span>)
<span class="kw">CorTestPlot</span>(dat)
<span class="kw">dev.off</span>()</code></pre></div>
<h3 id="pca-on-sd-data">PCA on sd data</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca.test &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">scale</span>(env.scale.correct.sd[[<span class="dv">6</span>]][ ,-<span class="dv">26</span>]))
<span class="kw">summary</span>(pca.test)
<span class="kw">rownames</span>(pca.test$rotation)

for (i in <span class="dv">1</span>:<span class="dv">6</span>){
    env.scale<span class="fl">.2</span>.sd.PC[[i]] &lt;-<span class="st"> </span><span class="kw">cbind</span>(env.scale.correct.sd[[i]],
                                    <span class="kw">scores</span>(<span class="kw">prcomp</span>(<span class="kw">scale</span>(env.scale.correct.sd[[i]][ ,-<span class="dv">26</span>])),
                                           <span class="dt">choices=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)))
}

<span class="kw">lapply</span>(env.scale<span class="fl">.2</span>.sd.PC, names)


<span class="kw">dim</span>(env.scale<span class="fl">.2</span>.sd.PC[[<span class="dv">4</span>]])

<span class="kw">tail</span>(env.scale<span class="fl">.2</span>.sd.PC[[<span class="dv">2</span>]][,<span class="dv">29</span>:<span class="dv">30</span>])
<span class="kw">summary</span>(env.scale<span class="fl">.2</span>.sd.PC[[<span class="dv">2</span>]][,<span class="dv">29</span>:<span class="dv">30</span>])

<span class="kw">filled.contour</span>((<span class="kw">matrix</span>(env.scale<span class="fl">.2</span>.sd.PC[[<span class="dv">2</span>]][,<span class="dv">30</span>], <span class="dt">ncol =</span> <span class="dv">25</span>, <span class="dt">byrow =</span> T)),
               <span class="dt">color.palette =</span> terrain.colors, <span class="dt">main =</span> <span class="st">&quot;slope.20m&quot;</span>)
                                        <span class="co"># ncol here corresponds to 500m / grain size (short side of bci)</span>

env.scale<span class="fl">.2</span>.sd.PC.correct &lt;-<span class="st"> </span>env.scale<span class="fl">.2</span>.sd.PC
<span class="kw">save</span>(env.scale<span class="fl">.2</span>.sd.PC.correct, <span class="dt">file =</span> <span class="st">&quot;env.scale.2.sd.PC.correct.Rdata&quot;</span>)

<span class="kw">lapply</span>(env.scale<span class="fl">.2</span>.sd.PC.correct, names)</code></pre></div>
<h3 id="generate-reduced-data-set-with-just-5-columns-1">Generate reduced data set with just 5 columns:</h3>
<ul>
<li>pH</li>
<li>elevation</li>
<li>shade</li>
<li>PC1</li>
<li>PC2</li>
</ul>
<p>The latter two based on all (but not shade) environmental variables, incl. the 4 soil types.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">env.scale.small.sd &lt;-<span class="st"> </span>env.scale<span class="fl">.2</span>.sd.PC.correct
<span class="kw">lapply</span>(env.scale.small.sd, names)

for (i in <span class="dv">1</span>:<span class="dv">6</span>){
    env.scale.small.sd[[i]] &lt;-<span class="st"> </span>env.scale<span class="fl">.2</span>.sd.PC.correct[[i]][,<span class="kw">c</span>(<span class="dv">24</span>,<span class="dv">8</span>,<span class="dv">26</span>,<span class="dv">29</span>,<span class="dv">30</span>)]
}

<span class="kw">lapply</span>(env.scale.small.sd, names)
<span class="kw">lapply</span>(env.scale.small.sd, str)

<span class="kw">save</span>(env.scale.small.sd, <span class="dt">file =</span> <span class="st">&quot;env.scale.small.sd.Rdata&quot;</span>)</code></pre></div>
<h3 id="match-demographic-rates-and-trait-data-from-joe-wright-with-species-data-all">Match demographic rates and trait data (from Joe Wright) with species data (all )</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">traits.new &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Data/Traits/</span>
<span class="st">Nadja_Traits/abun/new_fitness_measures_relpopchange.txt&quot;</span>, <span class="dt">stringsAsFactors =</span> F)
<span class="kw">dim</span>(traits.new)
<span class="kw">str</span>(traits.new)

<span class="kw">names</span>(traits.new)
<span class="kw">CorTestPlot</span>(traits.new[,-<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">7</span>)])</code></pre></div>
<p>Match to species data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index &lt;-<span class="st"> </span><span class="kw">which</span>(traits.new$species %in%<span class="st"> </span><span class="kw">colnames</span>(grid<span class="fl">.213.7</span>[[<span class="dv">7</span>]]))
traits213new &lt;-<span class="st"> </span>traits.new[index,]

<span class="kw">summary</span>(traits213new)

<span class="kw">colnames</span>(grid<span class="fl">.213.7</span>[[<span class="dv">7</span>]])
traits213new$species
<span class="kw">match</span>(traits213new$species, <span class="kw">colnames</span>(grid<span class="fl">.213.7</span>[[<span class="dv">7</span>]]))

traits213new2 &lt;-<span class="st"> </span>traits213new[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">7</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">8</span>:<span class="dv">12</span>)]
<span class="kw">rownames</span>(traits213new2) &lt;-<span class="st"> </span>traits213new2$species</code></pre></div>
<h3 id="joe-wrights-trait-data">Joe Wright's trait data:</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">traits.bci &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Data/Traits/</span>
<span class="st">Nadja_Traits/Traits_BCI.txt&quot;</span>)
<span class="kw">head</span>(traits.bci)
<span class="kw">dim</span>(traits.bci)
<span class="kw">match</span>(traits213new$species, traits.bci$species)
<span class="kw">summary</span>(traits.bci)

<span class="kw">library</span>(dplyr)
traits<span class="fl">.213</span>.all &lt;-<span class="st"> </span>dplyr:::<span class="kw">inner_join</span>(traits213new2, traits.bci, <span class="dt">by =</span> <span class="st">&quot;species&quot;</span>)
bci.data$traits<span class="fl">.213</span>.all &lt;-<span class="st"> </span>traits<span class="fl">.213</span>.all[,-<span class="dv">13</span>]
<span class="kw">rownames</span>(bci.data$traits<span class="fl">.213</span>.all) &lt;-<span class="st"> </span>bci.data$traits<span class="fl">.213</span>.all$species
<span class="kw">colnames</span>(bci.data$traits<span class="fl">.213</span>.all)[<span class="dv">1</span>] &lt;-<span class="st"> &quot;sp&quot;</span>

<span class="kw">save</span>(bci.data, <span class="dt">file =</span> <span class="st">&quot;bci.data.Rdata&quot;</span>)</code></pre></div>
<p>Complete trait data only are available for 164 out of the 213 species for which complete demographic rate data are available, mainly because data on seed mass for 46 species is missing.</p>
<p>Generate the trait data set based on the set of 164 species for which complete traits information (demographic rates and wright traits) are available:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">traits.wright &lt;-<span class="st"> </span>bci.data$traits<span class="fl">.213</span>.all[!<span class="kw">is.na</span>(<span class="kw">rowSums</span>(bci.data$traits<span class="fl">.213</span>.all[,<span class="kw">c</span>(<span class="dv">13</span>:<span class="dv">20</span>)])),]
bci.data$traits<span class="fl">.164</span>.all &lt;-<span class="st"> </span>traits.wright</code></pre></div>
<h1 id="estimating-species-interactions-using">Estimating species interactions using</h1>
<p>After having generated the environmental, abundance (and trait) data sets, species interactions were inferred using the Markov-network approach by Harris 2016 (Ecology). Because the analytical approach by Harris (2016) is intractable for large networks &gt;20 species, and does not account for environmental heterogeneity, I used the approximation approach to estimate interaction strength for the BCI-community containing 302 species. This approach also accounts for abiotic structure in the community data.</p>
<p>Although data on demographic rates and traits are not available for all BCI tree species, interactions were estimated based on the entire community (which in our case was 302 species) for each of the 7 spatial scales (grain sizes): 5m, 10m , 20m, 40m, 60m, 80m, 100m), and three size classes: 1-5cm (dbh), &gt;5-10cm (dbh), &gt;10cm (dbh):</p>
<h2 id="generating-abundance-grids">Generating abundance grids</h2>
<p>Although data on demographic rates and traits are not available for all BCI tree species, interactions were estimated based on the entire community (which in our case was 302 species) for each of the 7 spatial scales (grain sizes): 5m, 10m , 20m, 40m, 60m, 80m, 100m), and three size classes: 1-5cm (dbh), &gt;5-10cm (dbh), &gt;10cm (dbh):</p>
<h3 id="abundance-grids-for-302-species-across-scales">Abundance grids for 302 species, across scales</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path &lt;-<span class="st"> &quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Data/Traits/</span>
<span class="st">Nadja_Traits/abundance_gridcells/5m_Steps&quot;</span>

grid<span class="fl">.302.7</span> &lt;-<span class="st"> </span><span class="kw">list</span>()

for (j in <span class="kw">c</span>(<span class="st">&quot;gridcells5_1990&quot;</span>,<span class="st">&quot;gridcells10_1990&quot;</span>,<span class="st">&quot;gridcells20_1990&quot;</span>,<span class="st">&quot;gridcells40_1990&quot;</span>,
            <span class="st">&quot;gridcells60_1990&quot;</span>,<span class="st">&quot;gridcells80_1990&quot;</span>,<span class="st">&quot;gridcells100_1990&quot;</span>)){
    filenames &lt;-<span class="st"> </span><span class="kw">list.files</span>(path, <span class="dt">pattern=</span>j, <span class="dt">full.names=</span><span class="ot">TRUE</span>)
    ldf &lt;-<span class="st"> </span><span class="kw">lapply</span>(filenames, read.table)   
    
    grid<span class="fl">.302.7</span> &lt;-<span class="st"> </span><span class="kw">c</span>(grid<span class="fl">.302.7</span>, (ldf))
}

<span class="kw">lapply</span>(grid<span class="fl">.302.7</span>, dim)

<span class="kw">str</span>(grid<span class="fl">.302.7</span>)
<span class="kw">names</span>(grid<span class="fl">.302.7</span>) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;5m&quot;</span>,<span class="st">&quot;10m&quot;</span>,<span class="st">&quot;20m&quot;</span>,<span class="st">&quot;40m&quot;</span>,<span class="st">&quot;60m&quot;</span>,<span class="st">&quot;80m&quot;</span>,<span class="st">&quot;100m&quot;</span>)</code></pre></div>
<p>Prune the 305 species data sets to 302 species (40-80m scales):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grid<span class="fl">.302</span> &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span>:<span class="dv">7</span>) %do%<span class="st"> </span>{
    index &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">toupper</span>(<span class="kw">colnames</span>(grid<span class="fl">.302.7</span>[[i]])) %in%<span class="st"> </span><span class="kw">colnames</span>(grid<span class="fl">.302.7</span>[[<span class="dv">4</span>]]))
    grid<span class="fl">.302.7</span>[[i]] &lt;-<span class="st"> </span>grid<span class="fl">.302.7</span>[[i]][,index]
}
<span class="kw">lapply</span>(grid<span class="fl">.302</span>, dim)

<span class="kw">names</span>(grid<span class="fl">.302</span>) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;5m&quot;</span>,<span class="st">&quot;10m&quot;</span>,<span class="st">&quot;20m&quot;</span>,<span class="st">&quot;40m&quot;</span>,<span class="st">&quot;60m&quot;</span>,<span class="st">&quot;80m&quot;</span>,<span class="st">&quot;100m&quot;</span>)
<span class="kw">colnames</span>(grid<span class="fl">.302</span>[[<span class="dv">1</span>]]) &lt;-<span class="st"> </span><span class="kw">toupper</span>(<span class="kw">colnames</span>(grid<span class="fl">.302</span>[[<span class="dv">1</span>]]))

grid<span class="fl">.302</span> &lt;-<span class="st"> </span>grid<span class="fl">.302.7</span>
<span class="kw">save</span>(grid<span class="fl">.302.7</span>, <span class="dt">file =</span> <span class="st">&quot;grid.302.7.Rdata&quot;</span>)</code></pre></div>
<h3 id="abundance-grids-size-classes-across-scales-e.g.-10-cm-dbh">Abundance grids, size classes, across scales (e.g. &gt;10 cm dbh)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path &lt;-<span class="st"> &quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Data/Traits/</span>
<span class="st">Nadja_Traits/abundance_gridcells/SizeClasses&quot;</span>

grid.size<span class="fl">.10</span> &lt;-<span class="st"> </span><span class="kw">list</span>()

for (j in <span class="kw">c</span>(<span class="st">&quot;gridcells5_1990_10cm&quot;</span>,<span class="st">&quot;gridcells10_1990_10cm&quot;</span>,<span class="st">&quot;gridcells20_1990_10cm&quot;</span>,
            <span class="st">&quot;gridcells40_1990_10cm&quot;</span>,<span class="st">&quot;gridcells60_1990_10cm&quot;</span>,<span class="st">&quot;gridcells80_1990_10cm&quot;</span>,
            <span class="st">&quot;gridcells100_1990_10cm&quot;</span>)){
    filenames &lt;-<span class="st"> </span><span class="kw">list.files</span>(path, <span class="dt">pattern=</span>j, <span class="dt">full.names=</span><span class="ot">TRUE</span>)
    ldf &lt;-<span class="st"> </span><span class="kw">lapply</span>(filenames, read.table)   
    grid.size<span class="fl">.10</span> &lt;-<span class="st"> </span><span class="kw">c</span>(grid.size<span class="fl">.10</span>, (ldf))
}

<span class="kw">lapply</span>(grid.size<span class="fl">.10</span>, dim)
<span class="kw">names</span>(grid.size<span class="fl">.10</span>) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;5m&quot;</span>,<span class="st">&quot;10m&quot;</span>,<span class="st">&quot;20m&quot;</span>,<span class="st">&quot;40m&quot;</span>,<span class="st">&quot;60m&quot;</span>,<span class="st">&quot;80m&quot;</span>,<span class="st">&quot;100m&quot;</span>)

grid.full.size<span class="fl">.10</span> &lt;-<span class="st"> </span>grid.size<span class="fl">.10</span>
<span class="kw">save</span>(grid.full.size<span class="fl">.10</span>, <span class="dt">file =</span> <span class="st">&quot;grid.full.size.10.Rdata&quot;</span>)</code></pre></div>
<p>Repeat for size classes <code>&gt;5-10cm (dbh)</code> and <code>&gt;10cm (dbh)</code>.</p>
<p>Because the estimation of interactions using markov networks is computationally very expensive, particularily for the small spatial scales, the respective calculations were carried out on the EVE HPC, using the following environmental and species data sets:</p>
<p>Environmental data:</p>
<ul>
<li>env.scale.small.mean</li>
<li>env.scale.small.sd</li>
</ul>
<p>Species abundance grids:</p>
<ul>
<li>grid.full.302.7 (302 species)</li>
<li>grid.full.size.10 (228 species)</li>
<li>grid.full.size.5 (241 species)</li>
<li>grid.full.size.1 (281 species)</li>
</ul>
<h3 id="stochastic-approximation-of-markov-network-models">Stochastic approximation of markov network models</h3>
<p>Prior to analyses, I got the following email feedback from Dave Harris (11th Dec. 2015):</p>
<p>&quot;Hi Oliver, thanks for your interest and your question. If I understand correctly, you should be able to do it with a few minor modifications to my code.</p>
<ul>
<li>Run the first chunk with no modifications. You might need to install the mistnet package for the <code>%plus%</code> function. The easiest way to do that is with the <code>devtools</code> package: devtools::install_github(&quot;davharris/mistnet&quot;)</li>
<li>load your &quot;x&quot; and &quot;y&quot; data matrices and set n_loc, n_spp, and n_env to match your data (n_loc = nrow(y); n_spp = ncol(y); n_env = ncol(x))</li>
<li>Skip the second and third chunks, since they involve the generation of a simulated y matrix and you already have all your data.</li>
<li>Run the fourth chunk (&quot;Calculate sufficient statistics of the data&quot;) without modification</li>
<li>Run the fifth chunk (priors) without modification unless you have prior knowledge about the alpha and beta coefficients</li>
<li>Remove the lines about r-squared from the next chunk (R-squared can't be calculated unless the &quot;true&quot; coefficients are known) and then run the remaining code. This might take a while, depending on how many iterations you choose and how big your data set is.</li>
</ul>
<p>When you're done, you should have:</p>
<ul>
<li>an object called &quot;alpha_species&quot;: this is a vector of intercept terms for each species.</li>
<li>an object called &quot;alpha_env&quot;: this is a matrix of species' responses to the environmental variables (one row per variable, one column per species)</li>
<li>an object called &quot;beta&quot;: this is a symmetric matrix with each species' responses to one another.</li>
</ul>
<p>I don't think these objects will include names for the species or environmental variables, but they'll match the ordering of the columns in x and y. So the first row of alpha_env will refer to the first column of x, and so on.</p>
<p>Depending on the properties of your data set, the 50,000 iterations might not be enough for convergence. I'd try running it with 25,000 and 100,000 to make sure you get similar answers.</p>
<p>Hope this is useful for you! Let me know if you have any more questions or if you run into any cryptic errors.</p>
<p>I hope you have a great holiday. Say hi to Jon for me.</p>
<p>Dave&quot;</p>
<h4 id="general-code-for-the-stochastic-approximation-of-markov-network-models-incl.-daves-comments-above">General code for the stochastic approximation of markov network models (incl. Dave's comments above)</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rosalia)
<span class="kw">library</span>(mistnet)
<span class="kw">library</span>(vegan)</code></pre></div>
<p>Steps according to email by Dave Harris, 11th Dec. 2016:</p>
<p>Run the first chunk with no modifications.</p>
<ol>
<li>convenience function for adding intercepts to each column:</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="st">`</span><span class="dt">%plus%</span><span class="st">`</span> =<span class="st"> </span>mistnet:::<span class="st">`</span><span class="dt">%plus%</span><span class="st">`</span>
logistic =<span class="st"> </span><span class="kw">binomial</span>()$linkinv <span class="co"># logistic inverse link</span></code></pre></div>
<p>Random bernoulli trial</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbern =<span class="st"> </span>function(p){<span class="kw">rbinom</span>(<span class="kw">length</span>(p), <span class="dt">size =</span> <span class="dv">1</span>, <span class="dt">prob =</span> p)}</code></pre></div>
<ol>
<li>load your &quot;x&quot; and &quot;y&quot; data matrices and set n_loc, n_spp, and n_env to match your data (n_loc = nrow(y); n_spp = ncol(y); n_env = ncol(x))</li>
</ol>
<p>start with 20m scale since i have the env. variables for that scale:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;bci.data.Rdata&quot;</span>)</code></pre></div>
<ol>
<li>y-matric (species data):</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(bci.data)
<span class="kw">dim</span>(bci.data$grid<span class="fl">.213</span>[[<span class="dv">4</span>]])</code></pre></div>
<p>1250 sites x 213 species</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(bci.data$grid<span class="fl">.213</span>)
y &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(bci.data$grid<span class="fl">.213</span>[[<span class="dv">4</span>]])</code></pre></div>
<p>Transform into presence-absence matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">decostand</span>(y, <span class="st">&quot;pa&quot;</span>)</code></pre></div>
<ol>
<li>x-matric (environmental data). As did not have the proper soil data at that point, I downloaded some data from David Zeleny's webpage: <a href="http://www.davidzeleny.net/anadat-r/doku.php/en:data:bci:script-soil" class="uri">http://www.davidzeleny.net/anadat-r/doku.php/en:data:bci:script-soil</a></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">soil20x20 &lt;-
<span class="kw">read.delim</span>(<span class="st">'http://www.davidzeleny.net/anadat-r/lib/exe/</span>
<span class="st">fetch.php?media=data:bci-soil-20x20.txt'</span>)
<span class="kw">names</span>(soil20x20[,<span class="kw">c</span>(<span class="dv">8</span>,<span class="dv">11</span>,<span class="dv">13</span>:<span class="dv">15</span>)])
soil<span class="fl">.13</span> &lt;-<span class="st"> </span>soil20x20[,<span class="kw">c</span>(<span class="dv">3</span>:<span class="dv">15</span>)]
env =<span class="st"> </span>true_env &lt;-<span class="st"> </span>x &lt;-<span class="st"> </span>soil<span class="fl">.13</span> &lt;-<span class="st"> </span><span class="kw">scale</span>(soil20x20[,<span class="kw">c</span>(<span class="dv">3</span>:<span class="dv">15</span>)])</code></pre></div>
<p>Set some parameters</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
n_spp =<span class="st"> </span><span class="kw">ncol</span>(y)   <span class="co"># number of species</span>
n_loc =<span class="st"> </span><span class="kw">nrow</span>(y)    <span class="co"># number of locations</span>
n_env =<span class="st"> </span><span class="kw">ncol</span>(x)     <span class="co"># number of environmental predictors</span>
n_gibbs =<span class="st"> </span><span class="dv">5000</span>  <span class="co"># number of Gibbs sampling iterations</span></code></pre></div>
<ol>
<li><p>Skip the second and third chunks, since they involve the generation of a simulated y matrix and you already have all your data.</p></li>
<li><p>Run the fourth chunk (&quot;Calculate sufficient statistics of the data&quot;) without modification</p></li>
</ol>
<p>Calculate sufficient statistics of the data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y_stats =<span class="st"> </span><span class="kw">crossprod</span>(y)
y_env_stats =<span class="st"> </span><span class="kw">t</span>(true_env) %*%<span class="st"> </span>y</code></pre></div>
<p>Initialize the simulated landscape for stochastic approximation</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y_sim =<span class="st"> </span><span class="kw">matrix</span>(<span class="fl">0.5</span>, <span class="dt">nrow =</span> <span class="kw">nrow</span>(y), <span class="dt">ncol =</span> <span class="kw">ncol</span>(y))</code></pre></div>
<p>In this example, the true state of the environment is known without error</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">env =<span class="st"> </span>true_env</code></pre></div>
<p>Initialize species' responses to environment at 0. Also initialize the delta (change in parameter values from the previous optimization iteration) to zero, since no optimization has occurred yet</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha_env =<span class="st"> </span>delta_alpha_env =<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> n_env, <span class="dt">ncol =</span> n_spp)</code></pre></div>
<p>Initialize species' intercepts to match observed occurrence rates plus a small amount of regularization</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha_species =<span class="st"> </span><span class="kw">qlogis</span>((<span class="kw">colSums</span>(y) +<span class="st"> </span><span class="dv">1</span>) /<span class="st"> </span>(<span class="kw">nrow</span>(y) +<span class="st"> </span><span class="dv">2</span>))</code></pre></div>
<p>Initialize the deltas for the intercepts to zero</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">delta_alpha_species =<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, n_spp)</code></pre></div>
<p>Initialize pairwise interactions and deltas to zero</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">beta =<span class="st"> </span>delta_beta =<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> n_spp, <span class="dt">ncol =</span> n_spp)</code></pre></div>
<p>Overall alpha depends on alpha_species and alpha_env. Will be filled in later, so can initialize it with zeros no delta alpha to initialize b/c alpha not optimized directly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha =<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> n_spp, <span class="dt">ncol =</span> n_spp) </code></pre></div>
<p>5.) Run the fifth chunk (priors) without modification unless you have prior knowledge about the alpha and beta coefficients.</p>
<p>Very weak priors on alpha terms, somewhat stronger on beta terms.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha_env_prior =<span class="st"> </span>rosalia::<span class="kw">make_logistic_prior</span>(<span class="dt">scale =</span> <span class="dv">2</span>)$log_grad 
alpha_species_prior =<span class="st"> </span>rosalia::<span class="kw">make_logistic_prior</span>(<span class="dt">scale =</span> <span class="dv">2</span>)$log_grad
beta_prior =<span class="st"> </span>rosalia::<span class="kw">make_logistic_prior</span>(<span class="dt">scale =</span> <span class="fl">0.5</span>)$log_grad</code></pre></div>
<ol>
<li>Remove the lines about r-squared from the next chunk (R-squared can't be calculated unless the &quot;true&quot; coefficients are known) and then run the remaining code. This might take a while, depending on how many iterations you choose and how big your data set is.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">initial_learning_rate =<span class="st"> </span><span class="dv">1</span> <span class="co"># step size at start of optimization</span>
maxit =<span class="st"> </span><span class="dv">25000</span>             <span class="co"># Number of rounds of optimization # try 50000 and 100000 as well</span>
start_time =<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">Sys.time</span>())</code></pre></div>
<p>Record the timing history in this vector</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">times =<span class="st"> </span><span class="kw">integer</span>(maxit)

for(i in <span class="dv">1</span>:maxit){
  ##############################
  <span class="co"># Gibbs sampling for predicted species composition</span>
  ##############################

  <span class="co"># Update alpha</span>
  alpha =<span class="st"> </span>env %*%<span class="st"> </span>alpha_env %plus%<span class="st"> </span>alpha_species

  <span class="co"># Sample entries in y_sim from their conditional </span>
  <span class="co"># distribution (Gibbs sampling)</span>
    for(j in <span class="kw">sample.int</span>(n_spp)){
    y_sim[,j] =<span class="st"> </span><span class="kw">rbern</span>(<span class="kw">logistic</span>(alpha[ , j] +<span class="st"> </span>y_sim %*%<span class="st"> </span>beta[ , j]))
  }</code></pre></div>
<p>Stochastic approximation for updating alpha and beta</p>
<p>Update learning rate and momentum</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">learning_rate =<span class="st"> </span>initial_learning_rate *<span class="st"> </span><span class="dv">1000</span> /<span class="st"> </span>(<span class="dv">998</span> +<span class="st"> </span><span class="dv">1</span> +<span class="st"> </span>i)
momentum =<span class="st"> </span>.<span class="dv">9</span> *<span class="st"> </span>(<span class="dv">1</span> -<span class="st"> </span><span class="dv">1</span>/(.<span class="dv">1</span> *<span class="st"> </span>i +<span class="st"> </span><span class="dv">2</span>))</code></pre></div>
<p>Calculate sufficient statistics</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  y_sim_stats =<span class="st"> </span><span class="kw">crossprod</span>(y_sim)
  y_sim_env_stats =<span class="st"> </span><span class="kw">t</span>(env) %*%<span class="st"> </span>y_sim</code></pre></div>
<p>Calculate the gradient with respect to alpha and beta. Gradients are differences in sufficient statistics plus prior gradients, all divided by the number of locations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  stats_difference =<span class="st"> </span>y_stats -<span class="st"> </span>y_sim_stats
  beta_grad =<span class="st"> </span>(stats_difference +<span class="st"> </span><span class="kw">beta_prior</span>(beta)) /<span class="st"> </span>n_loc

  alpha_species_grad =<span class="st"> </span>(
    <span class="kw">diag</span>(stats_difference) +<span class="st"> </span>
<span class="st">      </span><span class="kw">alpha_species_prior</span>(alpha_species)
  ) /<span class="st"> </span>n_loc
  <span class="kw">diag</span>(beta_grad) =<span class="st"> </span><span class="dv">0</span> <span class="co"># beta_ii is 0 by convention</span>
  y_env_difference =<span class="st"> </span>y_env_stats -<span class="st"> </span>y_sim_env_stats
  alpha_env_grad =<span class="st"> </span>(y_env_difference +<span class="st"> </span>
<span class="st">                      </span><span class="kw">alpha_env_prior</span>(alpha_env))  /<span class="st"> </span>n_loc</code></pre></div>
<p>Calculate parameter updates: gradient times learning rate plus momentum times delta.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  delta_beta =<span class="st"> </span>beta_grad *<span class="st"> </span>learning_rate +<span class="st"> </span>momentum *<span class="st"> </span>delta_beta
  delta_alpha_species =<span class="st"> </span>alpha_species_grad *<span class="st"> </span>learning_rate +<span class="st"> </span>
<span class="st">    </span>momentum  *<span class="st"> </span>delta_alpha_species
  delta_alpha_env =<span class="st"> </span>alpha_env_grad *<span class="st"> </span>learning_rate +
<span class="st">    </span>momentum  *<span class="st"> </span>delta_alpha_env</code></pre></div>
<p>Add the deltas to the previous parameter values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  beta =<span class="st"> </span>beta +<span class="st"> </span>delta_beta
  alpha_species =<span class="st"> </span>alpha_species +<span class="st"> </span>delta_alpha_species
  alpha_env =<span class="st"> </span>alpha_env +<span class="st"> </span>delta_alpha_env

  times[i] =<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">Sys.time</span>()) -<span class="st"> </span>start_time
<span class="er">}</span></code></pre></div>
<p>Running time for 25000 iterations: 206 min</p>
<p>Save results:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MarkovNet213_20_env13_25000 &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;alpha_species&quot;</span>=alpha_species, <span class="st">&quot;alpha_env&quot;</span>=alpha_env,
                                    <span class="st">&quot;beta&quot;</span>=beta)

<span class="kw">save</span>(MarkovNet213_20_env13_25000, <span class="dt">file =</span> <span class="st">&quot;MarkovNet213_20_env13_25000.Rdata&quot;</span>)</code></pre></div>
<h1 id="scripts-for-analysis-on-eve">Scripts for analysis on EVE</h1>
<h2 id="bci-markov-submit-wrapper-correct-full-long.sh"><code>bci-markov-submit-wrapper-correct-full-long.sh</code></h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/bash</span>

<span class="ot">ITERATIONS=$1</span>
<span class="ot">JOB_NAME_SUFFIX=$2</span>

<span class="kw">for</span> <span class="kw">grid</span> in /data/idiv_chase/OliverPurschke/bci-input/grid.full.*.Rdata <span class="kw">;</span> <span class="kw">do</span>
  <span class="kw">for</span> <span class="kw">env</span> in /data/idiv_chase/OliverPurschke/bci-input/env.*.Rdata <span class="kw">;</span> <span class="kw">do</span>
    <span class="kw">for</span> <span class="kw">columns</span> in <span class="st">&quot;1&quot;</span> <span class="st">&quot;2&quot;</span> <span class="st">&quot;3&quot;</span> <span class="st">&quot;c(1,2)&quot;</span> <span class="st">&quot;4&quot;</span> <span class="st">&quot;c(4,5)&quot;</span> <span class="kw">;</span> <span class="kw">do</span>

      <span class="kw">case</span> <span class="ot">$env</span><span class="kw"> in</span> 
    *.mean*<span class="kw">)</span>
          <span class="ot">lenind=</span>0
          <span class="kw">;;</span>

    *.sd*<span class="kw">)</span>
          <span class="ot">lenind=</span>1
          <span class="kw">;;</span>
      <span class="kw">esac</span>

      <span class="kw">qsub</span> -N bci.markov-<span class="ot">$JOB_NAME_SUFFIX</span> ~/bci/bci-markov-submit-correct-full-long.sh <span class="ot">$grid</span> <span class="ot">$env</span> <span class="ot">$columns</span> <span class="ot">$lenind</span> <span class="ot">$ITERATIONS</span>
    <span class="kw">done</span>
  <span class="kw">done</span>
<span class="kw">done</span></code></pre></div>
<p>To submit, use:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">bash</span> bci-markov-submit-wrapper-correct-full-long.sh 1000 analyse-1</code></pre></div>
<h2 id="bci-markov-submit-correct-full-long.sh"><code>bci-markov-submit-correct-full-long.sh</code></h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/bash</span>

<span class="co">#$ -N bci.markov</span>

<span class="co">#$ -o /work/$USER/$JOB_NAME-$JOB_ID.out</span>
<span class="co">#$ -e /work/$USER/$JOB_NAME-$JOB_ID.err</span>

<span class="co">#$ -l h_rt=400:00:00</span>
<span class="co">#$ -l h_vmem=2G</span>
<span class="co">##$ -l avx # avx is cpu feature that is only available in the newer generation of the hardware</span>

<span class="co">#$ -pe smp 2</span>

<span class="kw">export</span> <span class="ot">MC_CORES=${NSLOTS:-</span>1<span class="ot">}</span>

<span class="kw">module</span> load R

<span class="kw">Rscript</span> ~/bci/bci.markov.correct.full.r <span class="st">&quot;</span><span class="ot">$@</span><span class="st">&quot;</span> /work/purschke/bci-output/<span class="ot">$JOB_NAME</span>-<span class="ot">$JOB_ID</span>-<span class="ot">$(</span><span class="kw">basename</span> <span class="ot">$1</span> .Rdata<span class="ot">)</span>_<span class="ot">$(</span><span class="kw">basename</span> <span class="ot">$2</span> .Rdata<span class="ot">)</span>_<span class="ot">${3}</span>_<span class="ot">${5}</span>.Rdata</code></pre></div>
<h2 id="bci.markov.correct.full.r"><code>bci.markov.correct.full.r</code></h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">args &lt;-<span class="st"> </span><span class="kw">commandArgs</span>(<span class="dt">trailingOnly =</span> <span class="ot">TRUE</span>)
grid &lt;-<span class="st"> </span>args[<span class="dv">1</span>] <span class="co"># e.g. &quot;grid.213.7.Rdata&quot;</span>
env.name &lt;-<span class="st"> </span>args[<span class="dv">2</span>] <span class="co"># e.g. &quot;env.scale.2.sd.PC.Rdata&quot;</span>
columns &lt;-<span class="st"> </span>args[<span class="dv">3</span>] <span class="co"># Arguments have type character, so coerce to numeric</span>
<span class="co"># see https://wiki.csiro.au/display/ASC/Run+R+script+as+a+batch+job+on+a+Linux+cluster</span>
len.ind &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(args[<span class="dv">4</span>]) <span class="co"># length index: 0 if env. data has all scales; 1 if the smallest scale is missing </span>
iters &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(args[<span class="dv">5</span>]) <span class="co"># markov iterations, e.g. 25000</span>
outpath &lt;-<span class="st"> </span>args[<span class="dv">6</span>]

### for an example:
## DO NOT RUN:
<span class="co"># 1) grid &lt;- &quot;grid.213.7.Rdata&quot;</span>
<span class="co"># 2) env.name &lt;- &quot;env.scale.2.sd.PC.Rdata&quot;</span>
<span class="co"># 3) columns &lt;- &quot;-c(26,29,30)&quot;</span>
<span class="co"># 4) len.ind &lt;- as.numeric(&quot;1&quot;)</span>
<span class="co"># 5) iters &lt;- as.numeric(&quot;10&quot;)</span>

<span class="co"># s &lt;- 5</span>
####

<span class="kw">library</span>(vegan)
<span class="kw">library</span>(rosalia)
<span class="kw">library</span>(mistnet)
<span class="kw">library</span>(doParallel)
<span class="co"># Find out how many cores are available (if you don't already know)</span>
<span class="kw">detectCores</span>()
<span class="co"># Create cluster with desired number of cores</span>
cl &lt;-<span class="st"> </span><span class="kw">makeForkCluster</span>() <span class="co"># !!! set to 2, since the first process (20,000 sites)takes ages on one core, while the other can finish on the second core </span>
<span class="co"># Register cluster</span>
<span class="kw">registerDoParallel</span>(cl)
<span class="co"># Find out how many cores are being used</span>
<span class="kw">getDoParWorkers</span>()

########
<span class="co"># # outside the loop</span>
<span class="st">`</span><span class="dt">%plus%</span><span class="st">`</span> =<span class="st"> </span>mistnet:::<span class="st">`</span><span class="dt">%plus%</span><span class="st">`</span>
logistic =<span class="st"> </span><span class="kw">binomial</span>()$linkinv <span class="co"># logistic inverse link</span>
rbern =<span class="st"> </span>function(p){<span class="kw">rbinom</span>(<span class="kw">length</span>(p), <span class="dt">size =</span> <span class="dv">1</span>, <span class="dt">prob =</span> p)}

<span class="co"># load(&quot;env.scale.2.sd.PC.Rdata&quot;) ### !!! change to &quot;mean&quot;</span>
env2 &lt;-<span class="st"> </span><span class="kw">get</span>(<span class="kw">load</span>(env.name))
<span class="co"># load(&quot;grid.213.7.Rdata&quot;) # !!! change to &quot;size&quot;</span>
grid2 &lt;-<span class="st"> </span><span class="kw">get</span>(<span class="kw">load</span>(grid))

######################
out &lt;-<span class="st"> </span><span class="kw">foreach</span> (<span class="dt">s =</span> <span class="dv">1</span>:<span class="kw">length</span>(env2), <span class="dt">.packages=</span><span class="kw">c</span>(<span class="st">&quot;rosalia&quot;</span>, <span class="st">&quot;mistnet&quot;</span>,<span class="st">&quot;vegan&quot;</span>)) %dopar%<span class="st"> </span>{ 
<span class="co"># !!! change to &quot;1:7&quot; if means are used; 1:6 if sd is used</span>
y &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(grid2[[s+len.ind]]) <span class="co"># !!! for &quot;sd&quot; change to [[s+1]]...</span>
<span class="co"># y &lt;- as.matrix(grid.213.7[[s]])                                     # !!! change to just &quot;s&quot;, if means instead of sd are used</span>
<span class="co"># transform into presence-absence matrix:</span>
y &lt;-<span class="st"> </span><span class="kw">decostand</span>(y, <span class="st">&quot;pa&quot;</span>)

###################################################################
####### to be changed depending on which env. data should be used
env =<span class="st"> </span>true_env &lt;-<span class="st"> </span>x &lt;-<span class="st"> </span>soil<span class="fl">.13</span> &lt;-<span class="st"> </span><span class="kw">scale</span>(env2[[s]][,<span class="kw">eval</span>(<span class="kw">parse</span>(<span class="dt">text=</span>columns))]) <span class="co">#!!!! change to the right variable set</span>
    <span class="co"># 1 pH</span>
    <span class="co"># 2 elevation</span>
    <span class="co"># 3 shade</span>
    <span class="co"># c(1,2) pH &amp; elevation </span>
    <span class="co"># 4 PC1</span>
    <span class="co"># c(4,5) PC1 &amp; PC2</span>
###################################################################

<span class="kw">set.seed</span>(<span class="dv">1</span>)
n_spp =<span class="st"> </span><span class="kw">ncol</span>(y)   <span class="co"># number of species</span>
n_loc =<span class="st"> </span><span class="kw">nrow</span>(y)    <span class="co"># number of locations</span>
n_env =<span class="st"> </span><span class="kw">ncol</span>(x)     <span class="co"># number of environmental predictors</span>

n_gibbs =<span class="st"> </span><span class="dv">5000</span>  <span class="co"># number of Gibbs sampling iterations</span>
######################

y_stats =<span class="st"> </span><span class="kw">crossprod</span>(y)
y_env_stats =<span class="st"> </span><span class="kw">t</span>(true_env) %*%<span class="st"> </span>y
y_sim =<span class="st"> </span><span class="kw">matrix</span>(<span class="fl">0.5</span>, <span class="dt">nrow =</span> <span class="kw">nrow</span>(y), <span class="dt">ncol =</span> <span class="kw">ncol</span>(y))
env =<span class="st"> </span>true_env
alpha_env =<span class="st"> </span>delta_alpha_env =<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> n_env, <span class="dt">ncol =</span> n_spp)
alpha_species =<span class="st"> </span><span class="kw">qlogis</span>((<span class="kw">colSums</span>(y) +<span class="st"> </span><span class="dv">1</span>) /<span class="st"> </span>(<span class="kw">nrow</span>(y) +<span class="st"> </span><span class="dv">2</span>))
delta_alpha_species =<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, n_spp)
beta =<span class="st"> </span>delta_beta =<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> n_spp, <span class="dt">ncol =</span> n_spp)
alpha =<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> n_spp, <span class="dt">ncol =</span> n_spp) 

alpha_env_prior =<span class="st"> </span>rosalia::<span class="kw">make_logistic_prior</span>(<span class="dt">scale =</span> <span class="dv">2</span>)$log_grad 
alpha_species_prior =<span class="st"> </span>rosalia::<span class="kw">make_logistic_prior</span>(<span class="dt">scale =</span> <span class="dv">2</span>)$log_grad
beta_prior =<span class="st"> </span>rosalia::<span class="kw">make_logistic_prior</span>(<span class="dt">scale =</span> <span class="fl">0.5</span>)$log_grad

initial_learning_rate =<span class="st"> </span><span class="dv">1</span> <span class="co"># step size at start of optimization</span>

##############################################
### needs to be changed
maxit =<span class="st"> </span>iters        
<span class="co"># !!! 25000, Number of rounds of optimization # try 50000 and 100000 as well</span>
########################################
start_time =<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">Sys.time</span>())
times =<span class="st"> </span><span class="kw">integer</span>(maxit)
        for(i in <span class="dv">1</span>:maxit){
  ##############################
  <span class="co"># Gibbs sampling for predicted species composition</span>
  ##############################

  <span class="co"># Update alpha</span>
  alpha =<span class="st"> </span>env %*%<span class="st"> </span>alpha_env %plus%<span class="st"> </span>alpha_species

  <span class="co"># Sample entries in y_sim from their conditional </span>
  <span class="co"># distribution (Gibbs sampling)</span>
    for(j in <span class="kw">sample.int</span>(n_spp)){
    y_sim[,j] =<span class="st"> </span><span class="kw">rbern</span>(<span class="kw">logistic</span>(alpha[ , j] +<span class="st"> </span>y_sim %*%<span class="st"> </span>beta[ , j]))
  }

  ##############################
  <span class="co"># Stochastic approximation for updating alpha and beta</span>
  ##############################

  <span class="co"># Update learning rate and momentum</span>
 
    learning_rate =<span class="st"> </span>initial_learning_rate *<span class="st"> </span><span class="dv">1000</span> /<span class="st"> </span>(<span class="dv">998</span> +<span class="st"> </span><span class="dv">1</span> +<span class="st"> </span>i)
 momentum =<span class="st"> </span>.<span class="dv">9</span> *<span class="st"> </span>(<span class="dv">1</span> -<span class="st"> </span><span class="dv">1</span>/(.<span class="dv">1</span> *<span class="st"> </span>i +<span class="st"> </span><span class="dv">2</span>))

  <span class="co"># Calculate sufficient statistics</span>
  y_sim_stats =<span class="st"> </span><span class="kw">crossprod</span>(y_sim)
  y_sim_env_stats =<span class="st"> </span><span class="kw">t</span>(env) %*%<span class="st"> </span>y_sim


  stats_difference =<span class="st"> </span>y_stats -<span class="st"> </span>y_sim_stats
  beta_grad =<span class="st"> </span>(stats_difference +<span class="st"> </span><span class="kw">beta_prior</span>(beta)) /<span class="st"> </span>n_loc

  alpha_species_grad =<span class="st"> </span>(
    <span class="kw">diag</span>(stats_difference) +<span class="st"> </span>
<span class="st">      </span><span class="kw">alpha_species_prior</span>(alpha_species)
  ) /<span class="st"> </span>n_loc
  <span class="kw">diag</span>(beta_grad) =<span class="st"> </span><span class="dv">0</span> <span class="co"># beta_ii is 0 by convention</span>
  y_env_difference =<span class="st"> </span>y_env_stats -<span class="st"> </span>y_sim_env_stats
  alpha_env_grad =<span class="st"> </span>(y_env_difference +<span class="st"> </span>
<span class="st">                      </span><span class="kw">alpha_env_prior</span>(alpha_env))  /<span class="st"> </span>n_loc


  delta_beta =<span class="st"> </span>beta_grad *<span class="st"> </span>learning_rate +<span class="st"> </span>momentum *<span class="st"> </span>delta_beta
  delta_alpha_species =<span class="st"> </span>alpha_species_grad *<span class="st"> </span>learning_rate +<span class="st"> </span>
<span class="st">    </span>momentum  *<span class="st"> </span>delta_alpha_species
  delta_alpha_env =<span class="st"> </span>alpha_env_grad *<span class="st"> </span>learning_rate +
<span class="st">    </span>momentum  *<span class="st"> </span>delta_alpha_env

  beta =<span class="st"> </span>beta +<span class="st"> </span>delta_beta
  alpha_species =<span class="st"> </span>alpha_species +<span class="st"> </span>delta_alpha_species
  alpha_env =<span class="st"> </span>alpha_env +<span class="st"> </span>delta_alpha_env
  times[i] =<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">Sys.time</span>()) -<span class="st"> </span>start_time
}
        <span class="kw">list</span>(<span class="st">&quot;alpha_species&quot;</span>=alpha_species, <span class="st">&quot;alpha_env&quot;</span>=alpha_env, <span class="st">&quot;beta&quot;</span>=beta)

}

##  give name here
<span class="co">#save(out, file = paste(grid,env.name,columns,iters,&quot;Rdata&quot;, sep = &quot;.&quot;))</span>
<span class="kw">save</span>(out, <span class="dt">file =</span> outpath)</code></pre></div>
<h1 id="grand-analysis-function">Grand analysis function</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mn.path &lt;-<span class="st"> &quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Results/MarkovNetworks/&quot;</span>
traits &lt;-<span class="st"> </span>bci.data$traits<span class="fl">.164</span>.all
phylo &lt;-<span class="st"> </span>bci.data$phy<span class="fl">.164</span>
<span class="co"># only in some cases (traits.213.all): give short names as rownames:</span>
<span class="kw">rownames</span>(traits) &lt;-<span class="st"> </span>bci.data$traits<span class="fl">.164</span>.all$sp

i &lt;-<span class="st"> &quot;size.1_&quot;</span>
v &lt;-<span class="st"> &quot;(4,5)_&quot;</span>
s &lt;-<span class="st"> </span><span class="dv">1</span>
t &lt;-<span class="st">  &quot;1:3&quot;</span>

mn.list &lt;-<span class="st"> </span><span class="kw">list.files</span>(mn.path, <span class="dt">pattern=</span><span class="kw">glob2rx</span>(<span class="kw">paste0</span>(<span class="st">&quot;*grid.full&quot;</span>,<span class="st">&quot;*&quot;</span>,i,<span class="st">&quot;*mean*&quot;</span>,v,<span class="st">&quot;*100000*&quot;</span>)))
                                        <span class="co"># to be changed if sd or 100000 iterations are used</span>
mn.list

<span class="co"># start loop</span>
<span class="kw">library</span>(doParallel)
out &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="kw">c</span>(<span class="st">&quot;302.7&quot;</span>, <span class="st">&quot;size.1_&quot;</span>, <span class="st">&quot;size.5_&quot;</span>, <span class="st">&quot;size.10_&quot;</span>)) %:%
<span class="st">    </span><span class="kw">foreach</span>(<span class="dt">v =</span> <span class="kw">c</span>(<span class="st">&quot;_1_&quot;</span>, <span class="st">&quot;_2_&quot;</span>, <span class="st">&quot;_3_&quot;</span>, <span class="st">&quot;(1,2)_&quot;</span>, <span class="st">&quot;_4_&quot;</span>, <span class="st">&quot;(4,5)_&quot;</span>), <span class="dt">.combine =</span> rbind) %do%<span class="st"> </span>{
        
        mn.path &lt;-<span class="st"> &quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Results/</span>
<span class="st">MarkovNetworks/&quot;</span>
        traits &lt;-<span class="st"> </span>bci.data$traits<span class="fl">.164</span>.all
        phylo &lt;-<span class="st"> </span>bci.data$phy<span class="fl">.164</span>
<span class="co"># only in some cases (traits.164.all): give short names as rownames:</span>
        <span class="kw">rownames</span>(traits) &lt;-<span class="st"> </span>bci.data$traits<span class="fl">.164</span>.all$sp

        mn.list &lt;-<span class="st"> </span><span class="kw">list.files</span>(mn.path, <span class="dt">pattern=</span><span class="kw">glob2rx</span>(<span class="kw">paste0</span>(<span class="st">&quot;*grid.full&quot;</span>,<span class="st">&quot;*&quot;</span>,i,<span class="st">&quot;*mean*&quot;</span>,
                                                              v,<span class="st">&quot;*100000*&quot;</span>))) 
                                        <span class="co"># to be changed if sd or 100000 iterations are used</span>
        mn.list
        mn.list &lt;-<span class="st"> </span><span class="kw">get</span>(<span class="kw">load</span>(<span class="kw">paste0</span>(mn.path, mn.list)))
                                        <span class="co"># match cooccurence matrices to traits</span>
        coocc &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">s =</span> <span class="dv">1</span>:<span class="kw">length</span>(mn.list)) %do%<span class="st"> </span>{
            beta &lt;-<span class="st"> </span>mn.list[[s]]$beta
            <span class="kw">rownames</span>(beta) &lt;-<span class="st"> </span><span class="kw">toupper</span>(<span class="kw">rownames</span>(beta))
            <span class="kw">colnames</span>(beta) &lt;-<span class="st"> </span><span class="kw">toupper</span>(<span class="kw">colnames</span>(beta))
            index.beta &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(beta) %in%<span class="st"> </span><span class="kw">rownames</span>(traits))
            beta.small &lt;-<span class="st"> </span>beta[index.beta, index.beta]
            index.traits &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">rownames</span>(traits) %in%<span class="st"> </span><span class="kw">rownames</span>(beta.small))
            traits.small &lt;-<span class="st"> </span>traits[index.traits,]
            mat.dis &lt;-<span class="st"> </span><span class="kw">match.comm.dist</span>(<span class="dt">comm =</span> <span class="kw">t</span>(traits.small[,<span class="kw">c</span>(<span class="dv">3</span>:<span class="kw">dim</span>(traits.small)[<span class="dv">2</span>])]), <span class="dt">dis =</span> <span class="kw">as.dist</span>(beta.small))$dis
            }        
        <span class="kw">names</span>(coocc) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;5m&quot;</span>, <span class="st">&quot;10m&quot;</span>, <span class="st">&quot;20m&quot;</span>, <span class="st">&quot;40m&quot;</span>, <span class="st">&quot;60m&quot;</span>, <span class="st">&quot;80m&quot;</span>, <span class="st">&quot;100m&quot;</span>)
        
        ## to be deleted later on: some extra stuff for the veech analysis
        <span class="co">#traits &lt;- bci.data$traits.213.all[,c(3:dim(bci.data$traits.213.all)[2])]</span>
        <span class="co">#coocc &lt;- veech.ses</span>
        ######################
        
                                        <span class="co"># match traits to cooccurrence</span>
        beta &lt;-<span class="st"> </span>mn.list[[<span class="dv">1</span>]]$beta
        <span class="kw">rownames</span>(beta) &lt;-<span class="st"> </span><span class="kw">toupper</span>(<span class="kw">rownames</span>(beta))
        <span class="kw">colnames</span>(beta) &lt;-<span class="st"> </span><span class="kw">toupper</span>(<span class="kw">colnames</span>(beta))
        index.beta &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(beta) %in%<span class="st"> </span><span class="kw">rownames</span>(traits))
        beta.small &lt;-<span class="st"> </span>beta[index.beta, index.beta]
        index.traits &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">rownames</span>(traits) %in%<span class="st"> </span><span class="kw">rownames</span>(beta.small))
        traits.small &lt;-<span class="st"> </span>traits[index.traits,]
        traits &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">match.comm.dist</span>(<span class="dt">comm =</span> <span class="kw">t</span>(traits.small[,<span class="kw">c</span>(<span class="dv">3</span>:<span class="kw">dim</span>(traits.small)[<span class="dv">2</span>])]), <span class="dt">dis =</span> coocc[[<span class="dv">1</span>]])$comm)
        <span class="kw">dim</span>(traits)

#################################
<span class="co"># for the demographic rates and wright traits:</span>
#################################
<span class="co">#1 &quot;1:3&quot;: &quot;demo.rec.light.mean2&quot;    &quot;demo.growth.light.mean2&quot; &quot;demo.mort.light.mean2&quot;</span>
<span class="co">#2 &quot;4:5&quot;: &quot;demo.growth.mean2&quot;  &quot;demo.mort.mean2&quot;</span>
<span class="co">#3 &quot;7&quot;: &quot;fitnessLogratio&quot;</span>
<span class="co">#4 &quot;10&quot;: &quot;relpopchange&quot;</span>
<span class="co">#5 &quot;19&quot;: &quot;fit_gro.mor_ratio&quot;</span>
<span class="co">#6 &quot;20&quot;: &quot;Niche.PC1&quot;</span>
<span class="co">#7 &quot;23&quot;: &quot;Fit.PC1&quot;</span>
<span class="co">#8 11: &quot;WoodDensity&quot;</span>
<span class="co">#9 12: &quot;MAXHEIGHT&quot;              </span>
<span class="co">#10 13 &quot;SeedMass&quot;</span>
<span class="co">#11 14: &quot;LMA&quot;                    </span>
<span class="co">#12 15: &quot;LeafArea&quot;</span>
<span class="co">#13 25: &quot;NP.PC1&quot;</span>
<span class="co">#14 27: &quot;Wright.8.PC1&quot;</span>
        
        trait.dist &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="kw">c</span>(<span class="st">&quot;1:3&quot;</span>, <span class="st">&quot;4:5&quot;</span>, <span class="st">&quot;7&quot;</span>, <span class="st">&quot;20&quot;</span>, <span class="st">&quot;23&quot;</span>,   
                                    <span class="st">&quot;11&quot;</span>, <span class="st">&quot;12&quot;</span>, <span class="st">&quot;13&quot;</span>, <span class="st">&quot;14&quot;</span>, <span class="st">&quot;15&quot;</span>, <span class="st">&quot;25&quot;</span>, <span class="st">&quot;27&quot;</span>,
                                    <span class="st">&quot;27:34&quot;</span>,
                                    <span class="st">&quot;35&quot;</span>, <span class="st">&quot;36&quot;</span>,
                                    <span class="st">&quot;37&quot;</span>,
                                    <span class="st">&quot;37:44&quot;</span>)) %do%<span class="st"> </span>{
            <span class="kw">dist</span>(traits[,<span class="kw">eval</span>(<span class="kw">parse</span>(<span class="dt">text=</span>i))])    
        }
        <span class="co">#lapply(trait.dist, range)</span>
        <span class="kw">names</span>(trait.dist) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Niche&quot;</span>, <span class="st">&quot;Fitness&quot;</span>, <span class="st">&quot;Fit.Log.Ratio&quot;</span>, <span class="st">&quot;Niche.PC1&quot;</span>, <span class="st">&quot;Fit.PC1&quot;</span>,   
                               <span class="st">&quot;WoodDensity&quot;</span>, <span class="st">&quot;MAXHEIGHT&quot;</span>, <span class="st">&quot;SeedMass&quot;</span>, <span class="st">&quot;LMA&quot;</span>, <span class="st">&quot;LeafArea&quot;</span>, <span class="st">&quot;NP.PC1&quot;</span>, <span class="st">&quot;Wright.8.PC1&quot;</span>,
                               <span class="st">&quot;Wright.8.PC1.8&quot;</span>,
                               <span class="st">&quot;SeedMass.log&quot;</span>, <span class="st">&quot;LeafArea.log&quot;</span>,
                               <span class="st">&quot;wright.8.trans.PC1&quot;</span>,
                               <span class="st">&quot;wright.8.trans.PC1.8&quot;</span>)

#################################
<span class="co"># just for demo traits:</span>
#################################
<span class="co">#1 &quot;1:3&quot;: &quot;demo.rec.light.mean2&quot;    &quot;demo.growth.light.mean2&quot; &quot;demo.mort.light.mean2&quot;</span>
<span class="co">#2 &quot;4:5&quot;: &quot;demo.growth.mean2&quot;  &quot;demo.mort.mean2&quot;</span>
<span class="co">#3 &quot;7&quot;: &quot;fitnessLogratio&quot;</span>
<span class="co">#4 &quot;10&quot;: &quot;relpopchange&quot;</span>
<span class="co">#5 &quot;19&quot;: &quot;fit_gro.mor_ratio&quot;</span>
<span class="co">#6 &quot;20&quot;: &quot;Niche.PC1&quot;</span>
<span class="co">#7 &quot;23&quot;: &quot;Fit.PC1&quot;</span>

<span class="co">#        trait.dist &lt;- foreach(t = c(&quot;1:3&quot;, &quot;4:5&quot;, &quot;7&quot;, &quot;10&quot;, &quot;19&quot;, &quot;20&quot;, &quot;23&quot;)) %do% {</span>
<span class="co">#            dist(traits[,eval(parse(text=t))])    </span>
<span class="co">#        }</span>
        <span class="co">#lapply(trait.dist, range)</span>
<span class="co">#        names(trait.dist) &lt;- c(&quot;Niche&quot;, &quot;Fitness&quot;, &quot;Fit.Log.Ratio&quot;, &quot;Rel.pop.change&quot;, &quot;Fit.gro.mor.ratio&quot;, &quot;Niche.PC1&quot;, &quot;Fit.PC1&quot;)</span>

<span class="co"># scale the data for multivariate analysis:</span>
<span class="co">#trait.dist.scale &lt;- lapply(trait.dist, scale)</span>
<span class="co">#lapply(trait.dist.scale, mean)</span>
<span class="co">#lapply(trait.dist.scale, sd)</span>

<span class="co"># dis.1 &lt;- dist(traits[,1:3])</span>
<span class="co">#dis.2 &lt;- dist(traits[,4:5])</span>

<span class="co"># do names in coocc and trait distance matrices match up?:</span>
<span class="co">#match(attr(coocc[[1]], &quot;Labels&quot;), attr(trait.dist[[1]], &quot;Labels&quot;))</span>

#######################################
<span class="co"># add phylogenetic distance matrix</span>
#######################################

<span class="co"># to match traits with phylogeny, give full species names as rownames to trait data:</span>
<span class="co">#str(traits)</span>
<span class="co">#class(traits)</span>
        traits<span class="fl">.2</span> &lt;-<span class="st"> </span>traits
        traits<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(traits<span class="fl">.2</span>)
        traits<span class="fl">.2</span>$sp &lt;-<span class="st"> </span><span class="kw">rownames</span>(traits<span class="fl">.2</span>)

<span class="co">#names(bci.data$traits.164.all)</span>

<span class="co"># 17-17.30: match phylo with traits</span>
        <span class="kw">library</span>(dplyr)
        traits<span class="fl">.2</span>.match &lt;-<span class="st"> </span><span class="kw">inner_join</span>(traits<span class="fl">.2</span>, bci.data$traits<span class="fl">.164</span>.all[,<span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">2</span>)])
        <span class="kw">rownames</span>(traits<span class="fl">.2</span>.match) &lt;-<span class="st"> </span>traits<span class="fl">.2</span>.match$species

        traits &lt;-<span class="st"> </span>traits<span class="fl">.2</span>.match[, -<span class="kw">c</span>(<span class="dv">25</span>:<span class="dv">26</span>)]

        mat &lt;-<span class="st"> </span><span class="kw">match.phylo.data</span>(<span class="dt">phy =</span> bci.data$phy<span class="fl">.164</span>, <span class="dt">data =</span> traits)
        <span class="co">#mat$phy</span>
<span class="co">#</span>
        phy.mat &lt;-<span class="st"> </span><span class="kw">as.dist</span>(<span class="kw">cophenetic</span>(mat$phy))

        phy.sort &lt;-<span class="st"> </span><span class="kw">match.comm.dist</span>(<span class="dt">comm =</span> <span class="kw">t</span>(traits[,<span class="kw">c</span>(<span class="dv">3</span>:<span class="kw">dim</span>(traits)[<span class="dv">2</span>])]), <span class="dt">dis =</span> phy.mat)$dis

<span class="co">#match(attr(coocc[[1]], &quot;Labels&quot;), attr(phy.sort, &quot;Labels&quot;))</span>
<span class="co">#match(attr(trait.dist[[1]], &quot;Labels&quot;), attr(phy.sort, &quot;Labels&quot;))</span>

        trait.dist.phy &lt;-<span class="st"> </span>trait.dist
        trait.dist.phy$phylo &lt;-<span class="st"> </span>phy.sort

<span class="co">#trait.dist.phy.scale &lt;- trait.dist.scale</span>
<span class="co">#trait.dist.phy.scale$phylo &lt;- scale(phy.sort)</span>

<span class="co">#lapply(trait.dist.phy.scale, mean)</span>
<span class="co">#lapply(trait.dist.phy.scale, sd)</span>


## univariate regression analysis

        coocc &lt;-<span class="st"> </span>coocc
        traits &lt;-<span class="st"> </span>trait.dist.phy <span class="co"># for veech, just use &quot;trait.dist&quot; here</span>
        
        ####################################################
        ## the following chunk might be deleted later on
        <span class="co"># put transformation here:</span>
        <span class="co"># log-transform seed mass and leaf area, sqrt-transform the rest:</span>
        <span class="co">#names(traits)[c(8,10)]</span>
        
        for(i in <span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>:<span class="dv">18</span>)){
            traits[[i]] &lt;-<span class="st"> </span><span class="kw">sqrt</span>(traits[[i]])
        }
        
        for(i in <span class="kw">c</span>(<span class="dv">8</span>,<span class="dv">10</span>)){
            traits[[i]] &lt;-<span class="st"> </span><span class="kw">log</span>(traits[[i]]+<span class="fl">0.0000001</span>) <span class="co"># there are zero distances for seed mass</span>
        }
        
        #########################################

    <span class="co"># include the data matching cleaning and matching bits in here</span>
            modelcoef &lt;-<span class="st"> </span><span class="kw">foreach</span> (<span class="dt">s =</span> <span class="dv">1</span>:<span class="kw">length</span>(coocc)) %:%<span class="st"> </span>
<span class="st">                </span><span class="kw">foreach</span>(<span class="dt">t =</span> <span class="dv">1</span>:<span class="kw">length</span>(traits), <span class="dt">.combine =</span> rbind) %do%<span class="st"> </span>{
                    <span class="kw">round</span>(<span class="kw">summary</span>(<span class="kw">lm</span>(coocc[[s]] ~<span class="st"> </span>traits[[t]]))$coefficients[<span class="dv">2</span>,], <span class="dv">4</span>)
                    <span class="co">#rownames(res) &lt;- names(traits)</span>
          
            }
            <span class="co">#rownames(res) &lt;- names(traits)</span>
                
       <span class="kw">names</span>(modelcoef) &lt;-<span class="st"> </span><span class="kw">names</span>(coocc)
       <span class="kw">ldply</span>(modelcoef)
    }

out</code></pre></div>
<h2 id="assign-names-to-output-from-grand-analysis-function">Assign names to output from <code>grand analysis function</code></h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(out) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;302.7&quot;</span>, <span class="st">&quot;size.1&quot;</span>, <span class="st">&quot;size.5&quot;</span>, <span class="st">&quot;size.10&quot;</span>)
out<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">ldply</span>(out)

nam &lt;-<span class="st"> </span><span class="kw">expand.grid</span>(
    <span class="dt">trait.dist =</span> <span class="kw">c</span>(<span class="st">&quot;Niche&quot;</span>, <span class="st">&quot;Fitness&quot;</span>, <span class="st">&quot;Fit.Log.Ratio&quot;</span>, <span class="st">&quot;Niche.PC1&quot;</span>, <span class="st">&quot;Fit.PC1&quot;</span>,   
                               <span class="st">&quot;WoodDensity&quot;</span>, <span class="st">&quot;MAXHEIGHT&quot;</span>, <span class="st">&quot;SeedMass&quot;</span>, <span class="st">&quot;LMA&quot;</span>, <span class="st">&quot;LeafArea&quot;</span>, <span class="st">&quot;NP.PC1&quot;</span>, <span class="st">&quot;Wright.8.PC1&quot;</span>,
                               <span class="st">&quot;Wright.8.PC1.8&quot;</span>,
                               <span class="st">&quot;SeedMass.log&quot;</span>, <span class="st">&quot;LeafArea.log&quot;</span>,
                               <span class="st">&quot;wright.8.trans.PC1&quot;</span>,
                               <span class="st">&quot;wright.8.trans.PC1.8&quot;</span>,<span class="st">&quot;Phylo&quot;</span>),
    <span class="dt">scale =</span> <span class="kw">c</span>(<span class="st">&quot;5&quot;</span>, <span class="st">&quot;10&quot;</span>, <span class="st">&quot;20&quot;</span>, <span class="st">&quot;40&quot;</span>, <span class="st">&quot;60&quot;</span>, <span class="st">&quot;80&quot;</span>, <span class="st">&quot;100&quot;</span>),
    <span class="dt">env =</span> <span class="kw">c</span>(<span class="st">&quot;1&quot;</span>, <span class="st">&quot;2&quot;</span>, <span class="st">&quot;3&quot;</span>, <span class="st">&quot;1,2&quot;</span>, <span class="st">&quot;4&quot;</span>, <span class="st">&quot;4,5&quot;</span>),
    <span class="dt">size =</span> <span class="kw">c</span>(<span class="st">&quot;All&quot;</span>, <span class="st">&quot;Size.1&quot;</span>, <span class="st">&quot;Size.5&quot;</span>, <span class="st">&quot;Size.10&quot;</span>)
)

out<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">cbind</span>(nam, out<span class="fl">.2</span>)
<span class="kw">head</span>(out<span class="fl">.3</span>)
out<span class="fl">.4</span> &lt;-<span class="st"> </span>out<span class="fl">.3</span>[,-<span class="kw">c</span>(<span class="dv">5</span>:<span class="dv">7</span>)]

<span class="kw">str</span>(out<span class="fl">.4</span>)
<span class="kw">names</span>(out<span class="fl">.4</span>)[<span class="dv">5</span>] &lt;-<span class="st"> &quot;t.val&quot;</span></code></pre></div>
<h2 id="plot-correlation-between-pairwise-interactions-and-demographic-rate-and-trait-differences">Plot correlation between pairwise interactions and demographic rate (and trait) differences</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sub<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">subset</span>(out<span class="fl">.4</span>, trait.dist %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Niche&quot;</span>, <span class="st">&quot;Fit.Log.Ratio&quot;</span>, <span class="st">&quot;WoodDensity&quot;</span>, <span class="st">&quot;MAXHEIGHT&quot;</span>, <span class="st">&quot;SeedMass&quot;</span>))

sub<span class="fl">.6</span> &lt;-<span class="st"> </span><span class="kw">subset</span>(out<span class="fl">.4</span>, trait.dist %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;LMA&quot;</span>, <span class="st">&quot;LeafArea&quot;</span>, <span class="st">&quot;NP.PC1&quot;</span>, <span class="st">&quot;Wright.8.PC1&quot;</span>, <span class="st">&quot;Wright.8.PC1.8&quot;</span>))

sub<span class="fl">.7</span> &lt;-<span class="st"> </span><span class="kw">subset</span>(out<span class="fl">.4</span>, trait.dist %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;SeedMass.log&quot;</span>, <span class="st">&quot;LeafArea.log&quot;</span>, <span class="st">&quot;wright.8.trans.PC1&quot;</span>, <span class="st">&quot;wright.8.trans.PC1.8&quot;</span>, <span class="st">&quot;Phylo&quot;</span>))

<span class="kw">pdf</span>(<span class="st">&quot;Cor.Coocc.Dist.wright.7.trans.100000.pdf&quot;</span>, <span class="dt">width =</span> <span class="dv">10</span>, <span class="dt">height =</span> <span class="dv">11</span>)
<span class="kw">ggplot</span>(sub<span class="fl">.7</span>, <span class="kw">aes</span>(scale, t.val, <span class="dt">color =</span> trait.dist, <span class="dt">group =</span> trait.dist)) +<span class="st"> </span>
<span class="st">    </span><span class="co">#geom_point() + </span>
<span class="st">    </span><span class="kw">geom_line</span>() +
<span class="st">    </span><span class="kw">facet_grid</span>(env ~<span class="st"> </span>size, <span class="dt">margins=</span>F) +
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">y=</span><span class="st">&quot;Cor.Coocc.Dist&quot;</span>, <span class="dt">x=</span> <span class="st">&quot;Scale&quot;</span>) +<span class="st"> </span>
<span class="st">    </span><span class="co">#scale_colour_manual(values=gs.pal(8)) +</span>
<span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept=</span><span class="kw">c</span>(<span class="dv">0</span>), <span class="dt">color =</span> <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>), <span class="dt">linetype=</span><span class="st">&quot;dashed&quot;</span>) +
<span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept=</span><span class="kw">c</span>(-<span class="fl">1.96</span>, <span class="fl">1.96</span>), <span class="dt">color =</span> <span class="kw">c</span>(<span class="st">&quot;grey&quot;</span>), <span class="dt">linetype=</span><span class="st">&quot;dashed&quot;</span>) +
<span class="st">    </span><span class="kw">theme_bw</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">panel.border =</span> <span class="kw">element_blank</span>(), <span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(),
                       <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>(),
                       <span class="dt">axis.line =</span> <span class="kw">element_line</span>(<span class="dt">colour =</span> <span class="st">&quot;black&quot;</span>))
<span class="kw">dev.off</span>()</code></pre></div>
<h2 id="plot-scale-dependent-changes-in-interaction-strength">Plot scale dependent changes in interaction strength</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">inter &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="kw">c</span>(<span class="st">&quot;302.7&quot;</span>, <span class="st">&quot;size.1_&quot;</span>, <span class="st">&quot;size.5_&quot;</span>, <span class="st">&quot;size.10_&quot;</span>)) %:%
<span class="st">    </span><span class="kw">foreach</span>(<span class="dt">v =</span> <span class="kw">c</span>(<span class="st">&quot;_1_&quot;</span>, <span class="st">&quot;_2_&quot;</span>, <span class="st">&quot;_3_&quot;</span>, <span class="st">&quot;(1,2)_&quot;</span>, <span class="st">&quot;_4_&quot;</span>, <span class="st">&quot;(4,5)_&quot;</span>), <span class="dt">.combine =</span> rbind) %do%<span class="st"> </span>{
        
        mn.path &lt;-<span class="st"> &quot;/home/oliver/Dokumente/PhD/PostPhD/Projects/BCI_Coexistence/Results/MarkovNetworks/&quot;</span>
        mn.list &lt;-<span class="st"> </span><span class="kw">list.files</span>(mn.path, <span class="dt">pattern=</span><span class="kw">glob2rx</span>(<span class="kw">paste0</span>(<span class="st">&quot;*grid.full&quot;</span>,<span class="st">&quot;*&quot;</span>,i,<span class="st">&quot;*mean*&quot;</span>,v,<span class="st">&quot;*50000*&quot;</span>))) 
                                        <span class="co"># to be changed if sd or 100000 iterations are used</span>
        mn.list
        mn.list &lt;-<span class="st"> </span><span class="kw">get</span>(<span class="kw">load</span>(<span class="kw">paste0</span>(mn.path, mn.list))) 
        
        coocc &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">s =</span> <span class="dv">1</span>:<span class="kw">length</span>(mn.list)) %do%<span class="st"> </span>{
            <span class="kw">as.dist</span>(mn.list[[s]]$beta)
            }        
        <span class="kw">names</span>(coocc) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;5m&quot;</span>, <span class="st">&quot;10m&quot;</span>, <span class="st">&quot;20m&quot;</span>, <span class="st">&quot;40m&quot;</span>, <span class="st">&quot;60m&quot;</span>, <span class="st">&quot;80m&quot;</span>, <span class="st">&quot;100m&quot;</span>)
        <span class="kw">ldply</span>(<span class="kw">lapply</span>(coocc, sd))
        }

<span class="kw">names</span>(inter) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;All&quot;</span>, <span class="st">&quot;Size.1&quot;</span>, <span class="st">&quot;Size.5&quot;</span>, <span class="st">&quot;Size.10&quot;</span>)
inter<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">ldply</span>(inter)

nam.inter &lt;-<span class="st"> </span><span class="kw">expand.grid</span>(
    <span class="dt">scale =</span> <span class="kw">c</span>(<span class="st">&quot;5&quot;</span>, <span class="st">&quot;10&quot;</span>, <span class="st">&quot;20&quot;</span>, <span class="st">&quot;40&quot;</span>, <span class="st">&quot;60&quot;</span>, <span class="st">&quot;80&quot;</span>, <span class="st">&quot;100&quot;</span>),
    <span class="dt">env =</span> <span class="kw">c</span>(<span class="st">&quot;1&quot;</span>, <span class="st">&quot;2&quot;</span>, <span class="st">&quot;3&quot;</span>, <span class="st">&quot;1,2&quot;</span>, <span class="st">&quot;4&quot;</span>, <span class="st">&quot;4,5&quot;</span>),
    <span class="dt">size =</span> <span class="kw">c</span>(<span class="st">&quot;All&quot;</span>, <span class="st">&quot;Size.1&quot;</span>, <span class="st">&quot;Size.5&quot;</span>, <span class="st">&quot;Size.10&quot;</span>)
)

inter<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">cbind</span>(nam.inter, inter<span class="fl">.2</span>)
<span class="kw">head</span>(inter<span class="fl">.3</span>)

inter<span class="fl">.4</span> &lt;-<span class="st"> </span>inter<span class="fl">.3</span>[,-<span class="kw">c</span>(<span class="dv">4</span>)]
<span class="kw">names</span>(inter<span class="fl">.4</span>)[<span class="dv">4</span>] &lt;-<span class="st"> &quot;Beta&quot;</span>


<span class="kw">pdf</span>(<span class="st">&quot;Inter.scales.env.sd.pdf&quot;</span>, <span class="dt">width =</span> <span class="dv">6</span>, <span class="dt">height =</span> <span class="dv">5</span>)
<span class="kw">ggplot</span>(inter<span class="fl">.4</span>, <span class="kw">aes</span>(scale, Beta, <span class="dt">color =</span> env, <span class="dt">group =</span> env)) +<span class="st"> </span>
<span class="st">    </span><span class="co">#geom_point() + </span>
<span class="st">    </span><span class="kw">geom_line</span>() +
<span class="st">    </span><span class="kw">facet_wrap</span>(~size) +
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">y=</span><span class="st">&quot;Interaction strength (Beta)&quot;</span>, <span class="dt">x=</span> <span class="st">&quot;Scale&quot;</span>) +<span class="st"> </span>
<span class="st">    </span><span class="co">#scale_colour_manual(values=gs.pal(8)) +</span>
<span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept=</span><span class="kw">c</span>(<span class="dv">0</span>), <span class="dt">color =</span> <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>), <span class="dt">linetype=</span><span class="st">&quot;dashed&quot;</span>)
    <span class="co">#geom_hline(yintercept=c(-1.96, 1.96), color = c(&quot;grey&quot;), linetype=&quot;dashed&quot;)</span>
<span class="kw">dev.off</span>()</code></pre></div>
<h1 id="some-additional-figures">Some additional figures</h1>
<h2 id="contour-plots-of-cooccurrence-vs.-niche-and-fitness-differences-across-scales">Contour plots of cooccurrence vs. niche and fitness differences across scales</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bci.data$traits.abu<span class="fl">.213</span> &lt;-<span class="st"> </span>traits.abu<span class="fl">.213</span>

<span class="kw">colnames</span>(bci.data$traits.abu<span class="fl">.213</span>)

<span class="co"># for &quot;fitnessSMA&quot;</span>
fit.dist &lt;-<span class="st"> </span><span class="kw">dist</span>(bci.data$traits.abu<span class="fl">.213</span>[,<span class="dv">8</span>])

####### continue here ###!!1
<span class="co"># for &quot;logratio&quot;</span>
fit.dist &lt;-<span class="st"> </span><span class="kw">dist</span>(bci.data$traits.abu<span class="fl">.213</span>[,<span class="dv">9</span>])
<span class="kw">str</span>(fit.dist)

<span class="co"># resp &lt;- check.213.df # to be changed accordingly</span>
niche &lt;-<span class="st"> </span>bci.data$niche.dist<span class="fl">.213</span>.df[,<span class="dv">1</span>]
fitness &lt;-<span class="st"> </span>(<span class="kw">as.vector</span>(fit.dist))

<span class="kw">library</span>(doParallel)
<span class="co"># Find out how many cores are available (if you don't already know)</span>
<span class="kw">detectCores</span>()
<span class="co"># Create cluster with desired number of cores</span>
cl &lt;-<span class="st"> </span><span class="kw">makeCluster</span>(<span class="dv">3</span>)

<span class="co"># Register cluster</span>
<span class="kw">registerDoParallel</span>(cl)

<span class="co"># Find out how many cores are being used</span>
<span class="kw">getDoParWorkers</span>()


for (j in <span class="dv">1</span>:<span class="dv">7</span>){ <span class="co"># loop across cooc-response variable (7 variants)</span>
    coocc &lt;-<span class="st"> </span>bci.data$coocc<span class="fl">.213</span>[[j]]
    figlist &lt;-<span class="st"> </span><span class="kw">foreach</span> (<span class="dt">i =</span> <span class="dv">1</span>:<span class="dv">18</span>, <span class="dt">combine =</span> list, <span class="dt">.packages=</span><span class="st">'lattice'</span>) %dopar%<span class="st"> </span>{ <span class="co"># loop across scales (grain sizes)</span>
        resp &lt;-coocc[,i]
        phylodistance &lt;-<span class="st"> </span><span class="kw">loess</span>(resp ~<span class="st"> </span>niche*fitness, <span class="dt">span =</span> <span class="dv">1</span>, <span class="dt">degree =</span> <span class="dv">1</span>)
        n.marginal &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="kw">min</span>(niche), <span class="kw">max</span>(niche), <span class="dt">length.out =</span> <span class="dv">50</span>)
        f.marginal &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="kw">min</span>(fitness), <span class="kw">max</span>(fitness), <span class="dt">length.out =</span> <span class="dv">50</span>)
        nf.marginal &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">niche =</span> n.marginal, <span class="dt">fitness =</span> f.marginal)
        grid &lt;-<span class="st"> </span><span class="kw">expand.grid</span>(nf.marginal)
        grid[, <span class="st">&quot;fit&quot;</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">predict</span>(phylodistance, grid))
        fig &lt;-<span class="st"> </span>lattice::<span class="kw">contourplot</span>(fit ~<span class="st"> </span>niche *<span class="st"> </span>fitness, <span class="dt">data =</span> grid, <span class="dt">cuts =</span> <span class="dv">10</span>, <span class="dt">region =</span> <span class="ot">TRUE</span>, <span class="dt">labels =</span> F, <span class="dt">contour =</span> F, <span class="dt">xlab =</span> <span class="st">&quot;Light response differences&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Demographic differences&quot;</span>, <span class="dt">main =</span> <span class="kw">colnames</span>(coocc)[i]
            <span class="co">#,</span>
            <span class="co">#panel=function(...){</span>
            <span class="co">#panel.contourplot(...)</span>
            <span class="co">#grid.points(niche, fitness, pch=1, size=unit(.001, &quot;char&quot;))</span>
          <span class="co">#}</span>
           <span class="co">#,</span>
            <span class="co">#       panel = function(x,y,z,...){</span>
            <span class="co">#  panel.contourplot(x,y,z,...)</span>
            <span class="co">#  panel.abline(0,1,lwd=1,col=&quot;blue&quot;)</span>
                                        <span class="co">#      }</span>
                                    )   
    <span class="co">#fig</span>
    
<span class="co"># pdf(file = paste(paste(&quot;Check.213.&quot;, i, sep=&quot;&quot;), &quot;pdf&quot;, sep=&quot;.&quot;), width = 5, height = 4.6, pointsize=12)</span>
<span class="co">#jpeg(filename = paste(paste(&quot;Coocc.check.wo.rec.a.spat.&quot;, i, sep=&quot;&quot;), &quot;jpeg&quot;, sep=&quot;.&quot;), width = 240, height = 226, quality = 100)</span>
<span class="co">#plot(fig)</span>
<span class="co">#dev.off()</span>
                                        <span class="co">#list(fig)</span>
        fig
    }
    
    <span class="kw">pdf</span>(<span class="dt">file =</span> <span class="kw">paste</span>(<span class="kw">paste</span>(<span class="st">&quot;Coocc.213.fitLogRatio&quot;</span>, j, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>), <span class="st">&quot;pdf&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;.&quot;</span>), <span class="dt">width =</span> <span class="dv">41</span>, <span class="dt">height =</span> <span class="dv">8</span>, <span class="dt">pointsize=</span><span class="dv">12</span>)
    <span class="kw">grid.arrange</span>(figlist[[<span class="dv">1</span>]],figlist[[<span class="dv">2</span>]],figlist[[<span class="dv">3</span>]],figlist[[<span class="dv">4</span>]],figlist[[<span class="dv">5</span>]],figlist[[<span class="dv">6</span>]],figlist[[<span class="dv">7</span>]],figlist[[<span class="dv">8</span>]],figlist[[<span class="dv">9</span>]],figlist[[<span class="dv">10</span>]],figlist[[<span class="dv">11</span>]],figlist[[<span class="dv">12</span>]],figlist[[<span class="dv">13</span>]],figlist[[<span class="dv">14</span>]],figlist[[<span class="dv">15</span>]],figlist[[<span class="dv">16</span>]],figlist[[<span class="dv">17</span>]],figlist[[<span class="dv">18</span>]],<span class="dt">ncol=</span><span class="dv">9</span>, <span class="dt">nrow=</span><span class="dv">2</span>)
    <span class="kw">dev.off</span>()
}</code></pre></div>
<h2 id="plotting-correlation-between-cooccurrence-vs.-niche-and-fitness-differences-across-scales">Plotting correlation between cooccurrence vs. niche and fitness differences across scales</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">niche.fit.scaled &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>((<span class="kw">cbind</span>(niche, fitness)))

modelcoef &lt;-<span class="st"> </span><span class="ot">NA</span>

for (i in <span class="dv">1</span>:<span class="dv">7</span>){
resp.vec &lt;-<span class="st"> </span>bci.data$check<span class="fl">.205</span>.ses.1s[,i]
mod &lt;-<span class="st"> </span><span class="kw">lm</span>(resp.vec ~<span class="st"> </span>niche*fitness, <span class="dt">data =</span> niche.fit.scaled)                   
modelcoef &lt;-<span class="st"> </span><span class="kw">rbind</span>(modelcoef, <span class="kw">melt</span>(<span class="kw">summary</span>(mod)$coefficients))
}
modelcoef &lt;-<span class="st"> </span>modelcoef[-<span class="dv">1</span>,]
modelcoef$scale &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">seq</span>(<span class="dv">10</span>,<span class="dv">100</span>,<span class="dv">5</span>), <span class="dt">each=</span><span class="dv">16</span>)
<span class="kw">colnames</span>(modelcoef)[<span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">2</span>)] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;variable&quot;</span>, <span class="st">&quot;coefficient&quot;</span>)
<span class="kw">str</span>(modelcoef)

modelcoef$variable &lt;-<span class="st"> </span><span class="kw">as.character</span>(modelcoef$variable)
modelcoef$coefficient &lt;-<span class="st"> </span><span class="kw">as.character</span>(modelcoef$coefficient)

 ##

modelcoef.estim &lt;-<span class="st"> </span>modelcoef[modelcoef$coefficient==<span class="st">&quot;t value&quot;</span>,]
modelcoef.estim<span class="fl">.2</span> &lt;-
modelcoef.estim[modelcoef.estim$variable!=<span class="st">&quot;(Intercept)&quot;</span>,]
modelcoef.estim<span class="fl">.2</span> &lt;-
modelcoef.estim<span class="fl">.2</span>[modelcoef.estim<span class="fl">.2</span>$variable!=<span class="st">&quot;niche.vec:fit.vec&quot;</span>,]

modelcoef.estim<span class="fl">.3</span> &lt;-<span class="st"> </span>modelcoef.estim<span class="fl">.2</span>[,-<span class="dv">2</span>]
<span class="co"># scale &gt;15</span>
modelcoef.estim<span class="fl">.4</span> &lt;-<span class="st"> </span>modelcoef.estim<span class="fl">.3</span>[modelcoef.estim<span class="fl">.3</span>$scale &gt;<span class="st"> </span><span class="dv">15</span>, ]
modelcoef.estim<span class="fl">.4</span>$variable &lt;-<span class="st"> </span><span class="kw">as.factor</span>(modelcoef.estim<span class="fl">.4</span>$variable)
modelcoef.estim<span class="fl">.4</span>$value &lt;-<span class="st"> </span>-(modelcoef.estim<span class="fl">.4</span>$value)
modelcoef.estim<span class="fl">.5</span> &lt;-<span class="st"> </span>modelcoef.estim<span class="fl">.4</span>

<span class="co"># rerun for t-value</span>

<span class="co"># do the lattice plot</span>
<span class="co"># include interaction</span>
div &lt;-<span class="st"> </span><span class="kw">xyplot</span>(value ~<span class="st"> </span>scale, <span class="dt">groups=</span>variable, <span class="dt">data =</span>
modelcoef.estim<span class="fl">.4</span>, <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>, <span class="dt">lty =</span> <span class="kw">c</span>(<span class="dv">1</span>), <span class="dt">par.settings =</span>
<span class="kw">list</span>(<span class="dt">axis.line =</span> <span class="kw">list</span>(<span class="dt">col =</span> <span class="dv">0</span>)),<span class="dt">scales=</span><span class="kw">list</span>(<span class="dt">col=</span><span class="dv">1</span>,<span class="dt">tck=</span><span class="kw">c</span>(-<span class="dv">1</span>,<span class="dv">0</span>)),  <span class="co">#</span>
remove top and right axes
              <span class="dt">panel=</span>function(...){
                lims &lt;-<span class="st"> </span><span class="kw">current.panel.limits</span>()
                <span class="kw">panel.xyplot</span>(...)
                <span class="kw">panel.abline</span>(<span class="dt">h=</span>lims$ylim[<span class="dv">1</span>],<span class="dt">v=</span>lims$xlim[<span class="dv">1</span>], <span class="dt">lwd =</span> <span class="fl">4.5</span>)
                <span class="kw">panel.abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">lwd=</span><span class="dv">1</span>, <span class="dt">lty=</span><span class="dv">2</span>, <span class="dt">col=</span><span class="st">&quot;black&quot;</span>)
              },
              <span class="dt">layout.heights=</span><span class="kw">list</span>(<span class="dt">axis.xlab.padding =</span> <span class="dv">1</span>),
<span class="dt">lwd =</span> <span class="fl">2.5</span>, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;green&quot;</span>,<span class="st">&quot;black&quot;</span>), <span class="dt">xlab =</span> <span class="st">&quot;Scale (m)&quot;</span>, <span class="dt">ylab =</span>
<span class="st">&quot;Correlation&quot;</span>, <span class="dt">key=</span><span class="kw">list</span>(<span class="dt">space=</span><span class="st">&quot;inside&quot;</span>,  <span class="dt">between =</span> <span class="dv">1</span>, <span class="dt">padding.text =</span>
<span class="dv">2</span>, <span class="dt">just =</span> <span class="kw">c</span>(.<span class="dv">6</span>, .<span class="dv">5</span>), <span class="dt">columns =</span> <span class="dv">1</span>, <span class="dt">lines =</span> <span class="kw">list</span>(<span class="dt">lty =</span> <span class="kw">c</span>(<span class="dv">1</span>), <span class="dt">lwd =</span> <span class="fl">2.5</span>,
<span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;green&quot;</span>,<span class="st">&quot;black&quot;</span>)),<span class="dt">text =</span> <span class="kw">list</span>(<span class="kw">c</span>(<span class="st">&quot;Fitness diff.&quot;</span>,
<span class="st">&quot;Niche.diff&quot;</span>, <span class="st">&quot;Niche x Fitness&quot;</span>))))
<span class="kw">plot</span>(div)


<span class="kw">pdf</span>(<span class="dt">file =</span> <span class="st">&quot;Scales.t_value.inter.PCA1-3.coef.scale.pdf&quot;</span>,<span class="dt">width =</span> <span class="fl">4.2</span>, <span class="dt">height =</span> <span class="dv">4</span>, <span class="dt">pointsize=</span><span class="dv">12</span>)
<span class="co"># postscript(file = &quot;BCI.DivergenceSize.eps&quot;,width = 6, height = 6.5,</span>
paper =<span class="st"> &quot;special&quot;</span>, onefile =<span class="st"> </span><span class="ot">FALSE</span>, horizontal =<span class="st"> </span><span class="ot">FALSE</span>, pointsize=<span class="dv">12</span><span class="er">)</span>
<span class="kw">plot</span>(div)
<span class="kw">dev.off</span>()</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">include_graphics</span>(<span class="st">&quot;Scales.t_value.inter.PCA1-3.coef.scale.pdf&quot;</span>, <span class="dt">dpi =</span> <span class="ot">NA</span>)</code></pre></div>
<p><embed src="data:application/pdf;base64,JVBERi0xLjQKJYHigeOBz4HTXHIKMSAwIG9iago8PAovQ3JlYXRpb25EYXRlIChEOjIwMTUxMDIxMTEzODAwKQovTW9kRGF0ZSAoRDoyMDE1MTAyMTExMzgwMCkKL1RpdGxlIChSIEdyYXBoaWNzIE91dHB1dCkKL1Byb2R1Y2VyIChSIDMuMi4yKQovQ3JlYXRvciAoUikKPj4KZW5kb2JqCjIgMCBvYmoKPDwgL1R5cGUgL0NhdGFsb2cgL1BhZ2VzIDMgMCBSID4+CmVuZG9iago3IDAgb2JqCjw8IC9UeXBlIC9QYWdlIC9QYXJlbnQgMyAwIFIgL0NvbnRlbnRzIDggMCBSIC9SZXNvdXJjZXMgNCAwIFIgPj4KZW5kb2JqCjggMCBvYmoKPDwKL0xlbmd0aCA4ODcgL0ZpbHRlciAvRmxhdGVEZWNvZGUKPj4Kc3RyZWFtCnictVZNbxs5DL3Pr9AxPpQVqe9rg7ZAgC2wGwM9ND0UqYMmcFI0LrD78/dxJI9UN23WCfYwAz9zKJJPj5TYnBk2N+bb9Gd7Xq2nl7u/3r4yl7vJkrXWjO/d5d308o3AZ31lWKgZ6qtidpmcNLC+NSfnl5+2G3NxcnuxWpn1zfR6fWygYfUXQ1D2FDWQUIhzpNOv9/eb7afv11/vxkjDY3ymwsYHYmfEWeJghB0FZ+435r25O3RoOZ6eP5Dj+ek7/JuC+Xv68NFY83lic4bnZuI5vz+mGi0lYmtuTfAUckNbc96s7JB/Nzc42HOiMtgr7HbxjqQs9gbVfsxW2oOtnN8SdakEenjm9wXbPa+69u+dXaJYsDtzkup9jK/o/nD25KrvUYGlbm6hVAUo9lANz9rU6MiGJqFbU1GwlOumsKpJFnODgx0OpXR7hd0uULWPi73BwR49FdftFXb7szc9JPJFScTW/8je49SzhTjj4OyPcQ6ZvBuc4xHOYrO2RXfOxziHSGGsme2hZB4fHIum+EFNMeX8mKY4Czm0+pQCxQhtBE1rO+VMgs6XolLawslqrWw98YyrxnLRuQuItkWSKWrbAmIeQ86IgD8BA5NLBkv7oLDqL0LyWSG8IhvshMxWrCmiH2cPKJb1BwZCtAqrNuFiZ6skstlgPvkZogExADIqTAqDaDOXUNNoui2sK2ynfSfyAWuVl4KFpdNiRblcaEEWYgdaHErLnRbGXOTYeWFkFEsnhkumkDozYgMl7tSIoHru3Ci23MlRMVjb2RGHynKnRzw3Bio/4pFP6QT17ytDUgfnQpFAcyEMHB0qq3KUHSW/cITU9JDZU4SSOAwUMc9nzEKR+PnMWCjSFAbtsE8UfKeIfWyUV4o4ijKxUMSx1JQbRYyZkgYB6fc5doo45rp+o4ghRy+dIg6lUt4o4sha1kKRfs9VRo7c0GZj02JWJiyzH639AN2b23EK7YNANe+PASPaDeA3jKsu53Zbdji4D+8QPx07TxoRKB1qwQ74hH1GIyG7BrXyB861w276TwFcva3sA1T4iwBPqgD3tSQ9QIX7AM+/fmI6QDsCyXLSYf7h5M3197vNbmc+X19drQwG7AmtPpr12RNuor+PCQWkUmO+u778sllBuQg2x/1fAjJOzHpBmuOZf0yr9RfX3/5M/wKGkG58ZW5kc3RyZWFtCmVuZG9iagozIDAgb2JqCjw8IC9UeXBlIC9QYWdlcyAvS2lkcyBbIDcgMCBSIF0gL0NvdW50IDEgL01lZGlhQm94IFswIDAgMzAyIDI4OF0gPj4KZW5kb2JqCjQgMCBvYmoKPDwKL1Byb2NTZXQgWy9QREYgL1RleHRdCi9Gb250IDw8L0YyIDEwIDAgUiA+PgovRXh0R1N0YXRlIDw8ID4+Ci9Db2xvclNwYWNlIDw8IC9zUkdCIDUgMCBSID4+Cj4+CmVuZG9iago1IDAgb2JqClsvSUNDQmFzZWQgNiAwIFJdCmVuZG9iago2IDAgb2JqCjw8IC9BbHRlcm5hdGUgL0RldmljZVJHQiAvTiAzIC9MZW5ndGggMjU5NiAvRmlsdGVyIC9GbGF0ZURlY29kZSA+PgpzdHJlYW0KeJydlndUU9kWh8+9N71QkhCKlNBraFICSA29SJEuKjEJEErAkAAiNkRUcERRkaYIMijggKNDkbEiioUBUbHrBBlE1HFwFBuWSWStGd+8ee/Nm98f935rn73P3Wfvfda6AJD8gwXCTFgJgAyhWBTh58WIjYtnYAcBDPAAA2wA4HCzs0IW+EYCmQJ82IxsmRP4F726DiD5+yrTP4zBAP+flLlZIjEAUJiM5/L42VwZF8k4PVecJbdPyZi2NE3OMErOIlmCMlaTc/IsW3z2mWUPOfMyhDwZy3PO4mXw5Nwn4405Er6MkWAZF+cI+LkyviZjg3RJhkDGb+SxGXxONgAoktwu5nNTZGwtY5IoMoIt43kA4EjJX/DSL1jMzxPLD8XOzFouEiSniBkmXFOGjZMTi+HPz03ni8XMMA43jSPiMdiZGVkc4XIAZs/8WRR5bRmyIjvYODk4MG0tbb4o1H9d/JuS93aWXoR/7hlEH/jD9ld+mQ0AsKZltdn6h21pFQBd6wFQu/2HzWAvAIqyvnUOfXEeunxeUsTiLGcrq9zcXEsBn2spL+jv+p8Of0NffM9Svt3v5WF485M4knQxQ143bmZ6pkTEyM7icPkM5p+H+B8H/nUeFhH8JL6IL5RFRMumTCBMlrVbyBOIBZlChkD4n5r4D8P+pNm5lona+BHQllgCpSEaQH4eACgqESAJe2Qr0O99C8ZHA/nNi9GZmJ37z4L+fVe4TP7IFiR/jmNHRDK4ElHO7Jr8WgI0IABFQAPqQBvoAxPABLbAEbgAD+ADAkEoiARxYDHgghSQAUQgFxSAtaAYlIKtYCeoBnWgETSDNnAYdIFj4DQ4By6By2AE3AFSMA6egCnwCsxAEISFyBAVUod0IEPIHLKFWJAb5AMFQxFQHJQIJUNCSAIVQOugUqgcqobqoWboW+godBq6AA1Dt6BRaBL6FXoHIzAJpsFasBFsBbNgTzgIjoQXwcnwMjgfLoK3wJVwA3wQ7oRPw5fgEVgKP4GnEYAQETqiizARFsJGQpF4JAkRIauQEqQCaUDakB6kH7mKSJGnyFsUBkVFMVBMlAvKHxWF4qKWoVahNqOqUQdQnag+1FXUKGoK9RFNRmuizdHO6AB0LDoZnYsuRlegm9Ad6LPoEfQ4+hUGg6FjjDGOGH9MHCYVswKzGbMb0445hRnGjGGmsVisOtYc64oNxXKwYmwxtgp7EHsSewU7jn2DI+J0cLY4X1w8TogrxFXgWnAncFdwE7gZvBLeEO+MD8Xz8MvxZfhGfA9+CD+OnyEoE4wJroRIQiphLaGS0EY4S7hLeEEkEvWITsRwooC4hlhJPEQ8TxwlviVRSGYkNimBJCFtIe0nnSLdIr0gk8lGZA9yPFlM3kJuJp8h3ye/UaAqWCoEKPAUVivUKHQqXFF4pohXNFT0VFysmK9YoXhEcUjxqRJeyUiJrcRRWqVUo3RU6YbStDJV2UY5VDlDebNyi/IF5UcULMWI4kPhUYoo+yhnKGNUhKpPZVO51HXURupZ6jgNQzOmBdBSaaW0b2iDtCkVioqdSrRKnkqNynEVKR2hG9ED6On0Mvph+nX6O1UtVU9Vvuom1TbVK6qv1eaoeajx1UrU2tVG1N6pM9R91NPUt6l3qd/TQGmYaYRr5Grs0Tir8XQObY7LHO6ckjmH59zWhDXNNCM0V2ju0xzQnNbS1vLTytKq0jqj9VSbru2hnaq9Q/uE9qQOVcdNR6CzQ+ekzmOGCsOTkc6oZPQxpnQ1df11Jbr1uoO6M3rGelF6hXrtevf0Cfos/ST9Hfq9+lMGOgYhBgUGrQa3DfGGLMMUw12G/YavjYyNYow2GHUZPTJWMw4wzjduNb5rQjZxN1lm0mByzRRjyjJNM91tetkMNrM3SzGrMRsyh80dzAXmu82HLdAWThZCiwaLG0wS05OZw2xljlrSLYMtCy27LJ9ZGVjFW22z6rf6aG1vnW7daH3HhmITaFNo02Pzq62ZLde2xvbaXPJc37mr53bPfW5nbse322N3055qH2K/wb7X/oODo4PIoc1h0tHAMdGx1vEGi8YKY21mnXdCO3k5rXY65vTW2cFZ7HzY+RcXpkuaS4vLo3nG8/jzGueNueq5clzrXaVuDLdEt71uUnddd457g/sDD30PnkeTx4SnqWeq50HPZ17WXiKvDq/XbGf2SvYpb8Tbz7vEe9CH4hPlU+1z31fPN9m31XfKz95vhd8pf7R/kP82/xsBWgHcgOaAqUDHwJWBfUGkoAVB1UEPgs2CRcE9IXBIYMj2kLvzDecL53eFgtCA0O2h98KMw5aFfR+OCQ8Lrwl/GGETURDRv4C6YMmClgWvIr0iyyLvRJlESaJ6oxWjE6Kbo1/HeMeUx0hjrWJXxl6K04gTxHXHY+Oj45vipxf6LNy5cDzBPqE44foi40V5iy4s1licvvj4EsUlnCVHEtGJMYktie85oZwGzvTSgKW1S6e4bO4u7hOeB28Hb5Lvyi/nTyS5JpUnPUp2Td6ePJninlKR8lTAFlQLnqf6p9alvk4LTduf9ik9Jr09A5eRmHFUSBGmCfsytTPzMoezzLOKs6TLnJftXDYlChI1ZUPZi7K7xTTZz9SAxESyXjKa45ZTk/MmNzr3SJ5ynjBvYLnZ8k3LJ/J9879egVrBXdFboFuwtmB0pefK+lXQqqWrelfrry5aPb7Gb82BtYS1aWt/KLQuLC98uS5mXU+RVtGaorH1futbixWKRcU3NrhsqNuI2ijYOLhp7qaqTR9LeCUXS61LK0rfb+ZuvviVzVeVX33akrRlsMyhbM9WzFbh1uvb3LcdKFcuzy8f2x6yvXMHY0fJjpc7l+y8UGFXUbeLsEuyS1oZXNldZVC1tep9dUr1SI1XTXutZu2m2te7ebuv7PHY01anVVda926vYO/Ner/6zgajhop9mH05+x42Rjf2f836urlJo6m06cN+4X7pgYgDfc2Ozc0tmi1lrXCrpHXyYMLBy994f9Pdxmyrb6e3lx4ChySHHn+b+O31w0GHe4+wjrR9Z/hdbQe1o6QT6lzeOdWV0iXtjusePhp4tLfHpafje8vv9x/TPVZzXOV42QnCiaITn07mn5w+lXXq6enk02O9S3rvnIk9c60vvG/wbNDZ8+d8z53p9+w/ed71/LELzheOXmRd7LrkcKlzwH6g4wf7HzoGHQY7hxyHui87Xe4Znjd84or7ldNXva+euxZw7dLI/JHh61HXb95IuCG9ybv56Fb6ree3c27P3FlzF3235J7SvYr7mvcbfjT9sV3qID0+6j068GDBgztj3LEnP2X/9H686CH5YcWEzkTzI9tHxyZ9Jy8/Xvh4/EnWk5mnxT8r/1z7zOTZd794/DIwFTs1/lz0/NOvm1+ov9j/0u5l73TY9P1XGa9mXpe8UX9z4C3rbf+7mHcTM7nvse8rP5h+6PkY9PHup4xPn34D94Tz+2VuZHN0cmVhbQplbmRvYmoKOSAwIG9iago8PAovVHlwZSAvRW5jb2RpbmcgL0Jhc2VFbmNvZGluZyAvV2luQW5zaUVuY29kaW5nCi9EaWZmZXJlbmNlcyBbIDQ1L21pbnVzIDk2L3F1b3RlbGVmdAoxNDQvZG90bGVzc2kgL2dyYXZlIC9hY3V0ZSAvY2lyY3VtZmxleCAvdGlsZGUgL21hY3JvbiAvYnJldmUgL2RvdGFjY2VudAovZGllcmVzaXMgLy5ub3RkZWYgL3JpbmcgL2NlZGlsbGEgLy5ub3RkZWYgL2h1bmdhcnVtbGF1dCAvb2dvbmVrIC9jYXJvbiAvc3BhY2VdCj4+CmVuZG9iagoxMCAwIG9iago8PCAvVHlwZSAvRm9udCAvU3VidHlwZSAvVHlwZTEgL05hbWUgL0YyIC9CYXNlRm9udCAvSGVsdmV0aWNhCi9FbmNvZGluZyA5IDAgUiA+PgplbmRvYmoKeHJlZgowIDExCjAwMDAwMDAwMDAgNjU1MzUgZiAKMDAwMDAwMDAyMSAwMDAwMCBuIAowMDAwMDAwMTYzIDAwMDAwIG4gCjAwMDAwMDEyNTAgMDAwMDAgbiAKMDAwMDAwMTMzMyAwMDAwMCBuIAowMDAwMDAxNDQ1IDAwMDAwIG4gCjAwMDAwMDE0NzggMDAwMDAgbiAKMDAwMDAwMDIxMiAwMDAwMCBuIAowMDAwMDAwMjkyIDAwMDAwIG4gCjAwMDAwMDQxNzMgMDAwMDAgbiAKMDAwMDAwNDQzMCAwMDAwMCBuIAp0cmFpbGVyCjw8IC9TaXplIDExIC9JbmZvIDEgMCBSIC9Sb290IDIgMCBSID4+CnN0YXJ0eHJlZgo0NTI3CiUlRU9GCg=="></embed></p>
<h1 id="r-settings"><code>R</code>-settings</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.1 (2017-06-30)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.5 LTS
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libopenblas.so.0
## LAPACK: /usr/lib/lapack/liblapack.so.3.0
## 
## locale:
##  [1] LC_CTYPE=de_DE.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=de_DE.UTF-8    
##  [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=de_DE.UTF-8   
##  [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] rmarkdown_1.6     ggplot2_2.2.1     adephylo_1.1-10  
##  [4] ade4_1.7-8        picante_1.6-2     nlme_3.1-131     
##  [7] gridExtra_2.2.1   phytools_0.6-20   maps_3.2.0       
## [10] ape_4.1           mistnet_0.3.0     rosalia_0.1.0    
## [13] dplyr_0.7.2.9000  doParallel_1.0.10 iterators_1.0.8  
## [16] foreach_1.4.3     vegan_2.4-4       lattice_0.20-35  
## [19] permute_0.9-4     knitr_1.17       
## 
## loaded via a namespace (and not attached):
##  [1] bold_0.5.0                gmodels_2.16.2           
##  [3] progress_1.1.2            httr_1.3.1               
##  [5] rprojroot_1.2             numDeriv_2016.8-1        
##  [7] backports_1.1.0           tools_3.4.1              
##  [9] R6_2.2.2                  DBI_0.7                  
## [11] lazyeval_0.2.0            mgcv_1.8-19              
## [13] colorspace_1.3-2          sp_1.2-5                 
## [15] prettyunits_1.0.2         mnormt_1.5-5             
## [17] phangorn_2.2.0            curl_2.8.1               
## [19] compiler_3.4.1            animation_2.5            
## [21] expm_0.999-2              xml2_1.1.1               
## [23] scales_0.5.0              mvtnorm_1.0-6            
## [25] quadprog_1.5-5            stringr_1.2.0            
## [27] digest_0.6.12             pkgconfig_2.0.1          
## [29] htmltools_0.3.6           plotrix_3.6-6            
## [31] highr_0.6                 rlang_0.1.2              
## [33] shiny_1.0.5               bindr_0.1                
## [35] combinat_0.0-8            jsonlite_1.5             
## [37] gtools_3.5.0              spdep_0.6-15             
## [39] magrittr_1.5              Matrix_1.2-11            
## [41] Rcpp_0.12.12              munsell_0.4.3            
## [43] yaml_2.1.14               scatterplot3d_0.3-40     
## [45] stringi_1.1.5             clusterGeneration_1.3.4  
## [47] MASS_7.3-47               plyr_1.8.4               
## [49] grid_3.4.1                gdata_2.18.0             
## [51] adegenet_2.0.1            deldir_0.1-14            
## [53] rncl_0.8.2                splines_3.4.1            
## [55] igraph_1.1.2              uuid_0.1-2               
## [57] taxize_0.8.9              boot_1.3-20              
## [59] seqinr_3.4-5              reshape2_1.4.2           
## [61] codetools_0.2-15          fastmatch_1.1-0          
## [63] LearnBayes_2.15           crul_0.3.8               
## [65] XML_3.98-1.9              GPArotation_2014.11-1    
## [67] glue_1.1.1                evaluate_0.10.1          
## [69] RcppArmadillo_0.7.960.1.2 RNeXML_2.0.7             
## [71] msm_1.6.4                 data.table_1.10.4        
## [73] httpuv_1.3.5              gtable_0.2.0             
## [75] purrr_0.2.3               tidyr_0.7.1              
## [77] reshape_0.8.7             assertthat_0.2.0         
## [79] mime_0.5                  phylobase_0.8.4          
## [81] xtable_1.8-2              coda_0.19-1              
## [83] survival_2.41-3           tibble_1.3.4             
## [85] bindrcpp_0.2              cluster_2.0.6</code></pre>
<h1 id="references">References</h1>

</body>
</html>
